// apipost-runtime/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var apipostSend = require('apipost-send');
var smCrypto = require('sm-crypto');
var urljoins$1 = require('urljoins');
var async = require('async');
var formData = require('form-data');
var cliTable3 = require('cli-table3');
require('cookie');
var zlib = require('zlib');
var buffer = require('buffer/');
var lodash = require('lodash');
var chai = require('chai');
var json5 = require('json5');
var uuid = require('uuid');
var apipostMock = require('apipost-mock');
var cryptoJs = require('crypto-js');
var jsonpath = require('jsonpath');
var x2js = require('x2js');
var jquery = require('jquery');
var ajaxForNode = require('ajax-for-node');
var jsencryptNode = require('jsencrypt-node');
var moment = require('moment');
var dayjs = require('dayjs');
var vm2 = require('vm2');
var apipostInsideTools = require('apipost-inside-tools');
var stripJsonComments = require('strip-json-comments');
var jsonBigint = require('json-bigint');
var apipostTools = require('apipost-tools');
var checkValidCookie = require('check-valid-cookie');
var urlJoin = require('url-join');
var fs = require('fs');
var path = require('path');
var mysql2 = require('mysql2');
var mssql = require('mssql');
var json2Csv = require('json-2-csv');
var testdataToApipostJson = require('testdata-to-apipost-json');
var clickhouse = require('clickhouse');
var pg = require('pg');
var child_process = require('child_process');
var atomicSleep = require('atomic-sleep');
var url = require('url');
var artTemplate = require('art-template');
var chaiApipost = require('chai-apipost');

function _interopDefaultLegacy(e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var apipostSend__default = /*#__PURE__*/_interopDefaultLegacy(apipostSend);
var smCrypto__default = /*#__PURE__*/_interopDefaultLegacy(smCrypto);
var urljoins__default = /*#__PURE__*/_interopDefaultLegacy(urljoins$1);
var async__default = /*#__PURE__*/_interopDefaultLegacy(async);
var formData__default = /*#__PURE__*/_interopDefaultLegacy(formData);
var cliTable3__default = /*#__PURE__*/_interopDefaultLegacy(cliTable3);
var zlib__default = /*#__PURE__*/_interopDefaultLegacy(zlib);
var buffer__default = /*#__PURE__*/_interopDefaultLegacy(buffer);
var lodash__default = /*#__PURE__*/_interopDefaultLegacy(lodash);
var chai__default = /*#__PURE__*/_interopDefaultLegacy(chai);
var json5__default = /*#__PURE__*/_interopDefaultLegacy(json5);
var uuid__default = /*#__PURE__*/_interopDefaultLegacy(uuid);
var apipostMock__default = /*#__PURE__*/_interopDefaultLegacy(apipostMock);
var cryptoJs__default = /*#__PURE__*/_interopDefaultLegacy(cryptoJs);
var jsonpath__default = /*#__PURE__*/_interopDefaultLegacy(jsonpath);
var x2js__default = /*#__PURE__*/_interopDefaultLegacy(x2js);
var jquery__default = /*#__PURE__*/_interopDefaultLegacy(jquery);
var ajaxForNode__default = /*#__PURE__*/_interopDefaultLegacy(ajaxForNode);
var jsencryptNode__default = /*#__PURE__*/_interopDefaultLegacy(jsencryptNode);
var moment__default = /*#__PURE__*/_interopDefaultLegacy(moment);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);
var vm2__default = /*#__PURE__*/_interopDefaultLegacy(vm2);
var apipostInsideTools__default = /*#__PURE__*/_interopDefaultLegacy(apipostInsideTools);
var stripJsonComments__default = /*#__PURE__*/_interopDefaultLegacy(stripJsonComments);
var jsonBigint__default = /*#__PURE__*/_interopDefaultLegacy(jsonBigint);
var apipostTools__default = /*#__PURE__*/_interopDefaultLegacy(apipostTools);
var checkValidCookie__default = /*#__PURE__*/_interopDefaultLegacy(checkValidCookie);
var urlJoin__default = /*#__PURE__*/_interopDefaultLegacy(urlJoin);
var fs__default = /*#__PURE__*/_interopDefaultLegacy(fs);
var path__default = /*#__PURE__*/_interopDefaultLegacy(path);
var mysql2__default = /*#__PURE__*/_interopDefaultLegacy(mysql2);
var mssql__default = /*#__PURE__*/_interopDefaultLegacy(mssql);
var json2Csv__default = /*#__PURE__*/_interopDefaultLegacy(json2Csv);
var testdataToApipostJson__default = /*#__PURE__*/_interopDefaultLegacy(testdataToApipostJson);
var clickhouse__default = /*#__PURE__*/_interopDefaultLegacy(clickhouse);
var pg__default = /*#__PURE__*/_interopDefaultLegacy(pg);
var child_process__default = /*#__PURE__*/_interopDefaultLegacy(child_process);
var atomicSleep__default = /*#__PURE__*/_interopDefaultLegacy(atomicSleep);
var url__default = /*#__PURE__*/_interopDefaultLegacy(url);
var artTemplate__default = /*#__PURE__*/_interopDefaultLegacy(artTemplate);
var chaiApipost__default = /*#__PURE__*/_interopDefaultLegacy(chaiApipost);

const { isUndefined, isString, isArray } = lodash__default["default"];
const getCollectionServerId$1 = (collection_id, collection_list) => {

    // step1:建立索引
    const parents_data = {};

    if (isArray(collection_list)) {
        collection_list.forEach(data => {
            parents_data[data?.target_id] = data;
        });
    }

    //递归查找
    const digFind = (target_id) => {
        const result = 'default'; // 默认使用default
        const targetInfo = parents_data?.[target_id];
        if (isUndefined(targetInfo)) {
            return result;
        }
        if (isString(targetInfo?.server_id) && targetInfo.server_id?.length > 0) {
            return targetInfo.server_id;
        }
        return digFind(targetInfo.parent_id);
    };
    return digFind(collection_id);
};



var utils = {
    getCollectionServerId: getCollectionServerId$1
};

const sm2 = smCrypto__default["default"].sm2, // add module for 7.0.8
    sm3 = smCrypto__default["default"].sm3, // add module for 7.0.8
    sm4 = smCrypto__default["default"].sm4, // add module for 7.0.8
    urljoins = urljoins__default["default"].urljoins,
    Buffer = buffer__default["default"].Buffer,// for 7.0.13
    { ClickHouse } = clickhouse__default["default"], // for 7.0.13
    { pgClient } = pg__default["default"];
const { getCollectionServerId } = utils;

// cli console
const cliConsole = function (args) {
    try {
        if (process.stdin.isTTY) {
            const commandName = process.argv[2];
            if (commandName == 'run' && typeof args == 'string') {
                console.log(args);
            }
        }
    } catch (e) { }
};

const Collection = function ApipostCollection(definition, option = { iterationCount: 1, sleep: 0 }) {
    const { iterationCount, sleep } = option;

    const definitionTlp = {
        parent_id: '-1', // 单任务的父ID
        event_id: '0', // 单任务的ID
        iteration: 0, // 当前执行第几轮循环（iteration）
        iterationCount: 0, // 本次执行需要循环的总轮数
        iterationData: {}, // excel导入的测试数据变量
        target_id: '',  // 接口ID ，仅适用于 api或者request
        request: {}, // 请求参数 ，仅适用于 api或者request
        response: {}, // 响应参数 ，仅适用于 api或者request
        cookie: [], // 响应cookie ，仅适用于 api或者request
        assert: [],
    };

    (function createRuntimeList(r, parent_id = '0') {
        if (r instanceof Array && r.length > 0) {
            r.forEach((item) => {
                lodash__default["default"].assign(item, definitionTlp, {
                    enabled: typeof item.enabled === 'undefined' ? 1 : item.enabled,
                    sort: typeof item.sort === 'undefined' ? 1 : item.sort,
                    parent_id,
                    event_id: item.event_id ? item.event_id : uuid__default["default"].v4(),
                    test_id: item.test_id ? item.test_id : uuid__default["default"].v4(),
                    type: item.type,
                    temp_env: lodash__default["default"].isObject(item.temp_env) ? item.temp_env : {}, // for 多环境
                    target_id: ['request', 'api', 'sample'].indexOf(item.type) > -1 ? item.data.target_id : '',
                    condition: ['request', 'api', 'sample'].indexOf(item.type) > -1 ? {} : item.data,
                    request: ['request', 'api', 'sample'].indexOf(item.type) > -1 ? item.data : {},
                    info: ['request', 'api', 'sample'].indexOf(item.type) > -1 ? {
                        // requestUrl: item.data.url ? item.data.url : item.data.request.url,
                        // requestName: item.data.name ? item.data.name : (item.data.url ? item.data.url : item.data.request.url),
                        requestId: item.data.target_id,
                    } : {},
                });

                if ((lodash__default["default"].isArray(item.children) && item.children.length > 0)) {
                    createRuntimeList(item.children, item.event_id);
                }
            });
        }
    }(definition));

    // 构造一个执行对象
    Object.defineProperty(this, 'definition', {
        configurable: true,
        writable: true,
        value: [lodash__default["default"].assign(lodash__default["default"].cloneDeep(definitionTlp), {
            type: 'for',
            condition: {
                limit: iterationCount > 0 ? iterationCount : 1,
                sleep: sleep > 0 ? sleep : 0,
            },
            enabled: 1,
            RUNNER_TOTAL_COUNT: lodash__default["default"].size(lodash__default["default"].filter(definition, lodash__default["default"].matchesProperty('enabled', 1))) * (iterationCount > 0 ? iterationCount : 1),
            children: lodash__default["default"].cloneDeep(definition),
        })],
    });
};

/*
@emitRuntimeEvent:请求脚本参数
@enableUnSafeShell:是否允许执行不安全脚本
*/
const Runtime = function ApipostRuntime(emitRuntimeEvent, enableUnSafeShell = true) {
    // 当前流程总错误计数器
    let RUNNER_TOTAL_COUNT = 0, // 需要跑的总event分母
        RUNNER_ERROR_COUNT = 0,
        RUNNER_PROGRESS = 0,
        RUNNER_RESULT_LOG = {},
        RUNNER_STOP = {};

    if (typeof emitRuntimeEvent !== 'function') {
        emitRuntimeEvent = function () { };
    }

    // Apipost 沙盒
    const Sandbox = function ApipostSandbox() {
        // 内置变量
        const insideVariablesScope = {
            list: {}, // 常量
        };

        /**
         *  拓展mockjs， 定义一些内置 mock
         *  fix bug for 7.0.8
         */
        const _mockjsRandomExtend = {};

        // 重写 string
        _mockjsRandomExtend['string'] = function (pool, start, end) {
            let charSet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

            if (typeof pool == 'string') {
                charSet = apipostMock__default["default"].mock(pool);
            }

            if (typeof pool == 'string') {
                pool = apipostMock__default["default"].mock(pool);
                try {
                    if (!lodash__default["default"].isNaN(Number(start))) {
                        if (lodash__default["default"].isNaN(Number(end)) === false) {
                            return lodash__default["default"].sampleSize(pool, lodash__default["default"].random(start, end)).join('');
                        }
                        return lodash__default["default"].sampleSize(pool, start).join('');
                    }
                }
                catch (ex) {
                    console.log(ex);
                }
                return lodash__default["default"].sample(pool);
            }

            if (typeof pool == 'number') {

                if (typeof start == 'number') {
                    return lodash__default["default"].sampleSize(charSet, lodash__default["default"].random(pool, start)).join('');
                }
                return lodash__default["default"].sampleSize(charSet, pool).join('')
            }
        };

        new Array('telephone', 'phone', 'mobile').forEach(func => {
            _mockjsRandomExtend[func] = function () {
                return this.pick(['131', '132', '137', '188']) + apipostMock__default["default"].mock(/\d{8}/)
            };
        });
        new Array('username', 'user_name', 'nickname', 'nick_name').forEach(func => {
            _mockjsRandomExtend[func] = function () {
                return apipostMock__default["default"].mock(`@cname`)
            };
        });
        new Array('avatar', 'icon', 'img', 'photo', 'pic').forEach(func => {
            _mockjsRandomExtend[func] = function () {
                return apipostMock__default["default"].mock(`@image('400x400')`)
            };
        });

        new Array('description').forEach(func => {
            _mockjsRandomExtend[func] = function () {
                return apipostMock__default["default"].mock(`@cparagraph`)
            };
        });

        new Array('id', 'userid', 'user_id', 'articleid', 'article_id').forEach(func => {
            _mockjsRandomExtend[func] = function () {
                return apipostMock__default["default"].mock(`@integer(100, 1000)`)
            };
        });

        apipostMock__default["default"].Random.extend(_mockjsRandomExtend);

        new Array('natural', 'integer', 'float', 'character', 'range', 'date', 'time', 'datetime', 'now', 'guid', 'integeincrementr', 'url', 'protocol', 'domain', 'tld', 'email', 'ip', 'region', 'province', 'city', 'county', 'county', 'zip', 'first', 'last', 'name', 'cfirst', 'clast', 'cname', 'color', 'rgb', 'rgba', 'hsl', 'paragraph', 'cparagraph', 'sentence', 'csentence', 'word', 'cword', 'title', 'ctitle', 'username', 'user_name', 'nickname', 'nick_name', 'avatar', 'icon', 'img', 'photo', 'pic', 'description', 'id', 'userid', 'user_id', 'articleid', 'article_id').forEach((func) => {
            insideVariablesScope.list[`$${func}`] = apipostMock__default["default"].mock(`@${func}`);
        });

        new Array('phone', 'mobile', 'telephone').forEach((func) => {
            insideVariablesScope.list[`$${func}`] = ['131', '132', '137', '188'][lodash__default["default"].random(0, 3)] + apipostMock__default["default"].mock(/\d{8}/);
        });

        // 兼容 v3
        insideVariablesScope.list.$timestamp = (function () {
            return Date.parse(new Date()) / 1000;
        }());

        insideVariablesScope.list.$microTimestamp = (function () {
            return (new Date()).getTime();
        }());

        insideVariablesScope.list.$randomInt = (function () {
            return Math.floor(Math.random() * 1000);
        }());

        insideVariablesScope.list.$randomFloat = (function () {
            return Math.random() * 1000;
        }());

        // 动态变量
        const variablesScope = {
            globals: {}, // 公共变量
            environment: {}, // 环境变量
            collectionVariables: {}, // 目录变量 当前版本不支持，目前为兼容postman
            variables: {}, // 临时变量，无需存库
            iterationData: {}, // 流程测试时的数据变量，临时变量，无需存库
        };

        // 获取所有动态变量
        function getAllDynamicVariables(type) {
            if (typeof aptScripts === 'object') {
                Object.keys(variablesScope).forEach((key) => {
                    if (lodash__default["default"].isObject(aptScripts[key]) && lodash__default["default"].isFunction(aptScripts[key].toObject) && ['iterationData', 'variables'].indexOf(key) > -1) {
                        lodash__default["default"].assign(variablesScope[key], aptScripts[key].toObject());
                    }
                });
            }

            if (variablesScope.hasOwnProperty(type)) {
                return lodash__default["default"].isObject(variablesScope[type]) ? variablesScope[type] : {};
            }
            const allVariables = {};
            Object.keys(variablesScope).forEach((type) => {
                lodash__default["default"].assign(allVariables, variablesScope[type]);
            });

            return allVariables;
        }

        // 设置动态变量
        const dynamicVariables = {};

        // 变量相关
        // ['variables'] 临时变量
        Object.defineProperty(dynamicVariables, 'variables', {
            configurable: true,
            value: {
                set(key, value) {
                    if (lodash__default["default"].isObject(value)) {
                        try {
                            value = JSON.stringify(value);
                        } catch (e) {
                            value = String(value);
                        }
                    }
                    variablesScope.variables[key] = value;
                },
                get(key) {
                    const allVariables = getAllDynamicVariables();
                    return allVariables[key];
                },
                has(key) {
                    return getAllDynamicVariables().hasOwnProperty(key);
                },
                delete(key) {
                    delete variablesScope.variables[key];
                },
                unset(key) {
                    delete variablesScope.variables[key];
                },
                clear() {
                    if (lodash__default["default"].isObject(variablesScope.variables)) {
                        lodash__default["default"].forEach(variablesScope.variables, (value, key) => {
                            delete variablesScope.variables[key];
                        });
                    }
                    variablesScope.variables = {};
                },
                replaceIn(variablesStr) {
                    return replaceIn(variablesStr);
                },
                toObject() {
                    return getAllDynamicVariables();
                },
            },
        });

        // ['iterationData'] 临时变量
        Object.defineProperty(dynamicVariables, 'iterationData', {
            configurable: true,
            value: {
                set(key, value) {
                    variablesScope.iterationData[key] = value;
                },
                get(key) {
                    return variablesScope.iterationData[key];
                },
                has(key) {
                    return variablesScope.iterationData.hasOwnProperty(key);
                },
                replaceIn(variablesStr) {
                    return replaceIn(variablesStr, 'iterationData');
                },
                toObject() {
                    return variablesScope.iterationData;
                },
                clear() {
                    if (lodash__default["default"].isObject(variablesScope.iterationData)) {
                        lodash__default["default"].forEach(variablesScope.iterationData, (value, key) => {
                            delete variablesScope.iterationData[key];
                        });
                    }
                    variablesScope.iterationData = {};
                },
            },
        });

        // ['globals', 'environment', 'collectionVariables']
        Object.keys(variablesScope).forEach((type) => {
            if (['iterationData', 'variables'].indexOf(type) === -1) {
                Object.defineProperty(dynamicVariables, type, {
                    configurable: true,
                    value: {
                        set(key, value, emitdb = true) {
                            if (lodash__default["default"].isObject(value)) {
                                try {
                                    value = JSON.stringify(value);
                                } catch (e) {
                                    value = String(value);
                                }
                            }

                            variablesScope[type][key] = value;

                            if (emitdb) {
                                typeof aptScripts === 'object' && lodash__default["default"].isObject(aptScripts[type]) && lodash__default["default"].isFunction(aptScripts[type].set) && aptScripts[type].set(key, value);
                            }
                        },
                        get(key) {
                            return variablesScope[type][key];
                        },
                        has(key) {
                            return variablesScope[type].hasOwnProperty(key);
                        },
                        delete(key) {
                            delete variablesScope[type][key];
                            typeof aptScripts === 'object' && lodash__default["default"].isObject(aptScripts[type]) && lodash__default["default"].isFunction(aptScripts[type].delete) && aptScripts[type].delete(key);
                        },
                        unset(key) {
                            delete variablesScope[type][key];
                            typeof aptScripts === 'object' && lodash__default["default"].isObject(aptScripts[type]) && lodash__default["default"].isFunction(aptScripts[type].delete) && aptScripts[type].delete(key);
                        },
                        clear() {
                            if (lodash__default["default"].isObject(variablesScope[type])) { // fix bug
                                lodash__default["default"].forEach(variablesScope[type], (value, key) => {
                                    delete variablesScope[type][key];
                                });
                            }
                            variablesScope[type] = {};
                            typeof aptScripts === 'object' && lodash__default["default"].isObject(aptScripts[type]) && lodash__default["default"].isFunction(aptScripts[type].clear) && aptScripts[type].clear();
                        },
                        replaceIn(variablesStr) {
                            return replaceIn(variablesStr, type);
                        },
                        toObject() {
                            return variablesScope[type];
                        },
                    },
                });
            }
        });

        // 获取所有内置变量
        function getAllInsideVariables() {
            return lodash__default["default"].cloneDeep(insideVariablesScope.list);
        }

        // 变量替换
        function replaceIn(variablesStr, type, withMock = false) {
            if (!lodash__default["default"].isString(variablesStr)) { // fix bug
                return variablesStr;
            }

            let allVariables = getAllInsideVariables(); // fix bug
            // console.log(getAllDynamicVariables(type));
            // let allVariables = {};
            lodash__default["default"].assign(allVariables, getAllDynamicVariables(type));

            if (withMock) {
                try {
                    // console.log(variablesStr, typeof variablesStr, Mock.mock(`${variablesStr}`))
                    variablesStr = apipostMock__default["default"].mock(variablesStr);
                } catch (e) { console.log(e); }
            }

            // 替换自定义变量
            const _regExp = new RegExp(Object.keys(allVariables).map((item) => {
                if (lodash__default["default"].startsWith(item, '$')) {
                    item = `\\${item}`;
                }
                return `\\{\\{${item}\\}\\}`; // fix bug
            }).join('|'), 'gi');

            variablesStr = lodash__default["default"].replace(variablesStr, _regExp, (key) => {
                let reStr = allVariables[String(lodash__default["default"].replace(key, /[{}]/gi, ''))];
                if (lodash__default["default"].isString(reStr)) {
                    reStr = reStr.replace(/\n/g, '\\n');      //bugfix v7.1.4
                }
                // console.log(String(_.replace(key, /[{}]/gi, '')), reStr, _regExp);
                if (typeof reStr !== 'undefined') {
                    return reStr;
                }
                return key;
            });
            // console.log('allVariables', variablesStr, _regExp);
            allVariables = null;
            return variablesStr;
        }

        // console
        const consoleFn = {};
        new Array('log', 'warn', 'info', 'error').forEach((method) => {
            Object.defineProperty(consoleFn, method, {
                configurable: true,
                value() {
                    emitRuntimeEvent({
                        action: 'console',
                        method,
                        message: {
                            type: 'log',
                            data: Array.from(arguments),
                        },
                        timestamp: Date.now(),
                        datetime: dayjs__default["default"]().format('YYYY-MM-DD HH:mm:ss'),
                    });
                },
            });
        });

        // 发送断言结果
        function emitAssertResult(status, expect, result, scope) {
            if (typeof scope !== 'undefined' && lodash__default["default"].isObject(scope) && lodash__default["default"].isArray(scope.assert)) {
                // 更新日志
                const item = lodash__default["default"].isObject(RUNNER_RESULT_LOG) ? RUNNER_RESULT_LOG[scope.iteration_id] : {};

                if (item) {
                    if (!lodash__default["default"].isArray(item.assert)) {
                        item.assert = [];
                    }

                    item.assert.push({
                        status,
                        expect,
                        result,
                    });
                    // console.log(item, item.assert)
                    if (status === 'success') {
                        cliConsole('\t✓'.green + ` ${expect} 匹配`.grey);
                    } else {
                        RUNNER_ERROR_COUNT++;
                        cliConsole(`\t${RUNNER_ERROR_COUNT}. ${expect} ${result}`.bold.red);
                    }
                }
            }
        }

        // 设置响应和请求参数
        function emitTargetPara(data, scope) {
            if (typeof scope !== 'undefined' && lodash__default["default"].isObject(scope)) {
                // 更新日志

                if (lodash__default["default"].isObject(RUNNER_RESULT_LOG)) {
                    const item = RUNNER_RESULT_LOG[scope.iteration_id];

                    if (item) {
                        switch (data.action) {
                            case 'SCRIPT_ERROR':
                                if (item.type == 'api' || item.type == 'sample') {
                                    lodash__default["default"].set(item, `script_error.${data.eventName}`, data.data);
                                }
                                break;
                        }
                    }
                }
            }
        }

        // 发送可视化结果
        function emitVisualizerHtml(status, html, scope) {
            if (typeof scope !== 'undefined' && lodash__default["default"].isObject(scope)) {
                if (lodash__default["default"].isObject(RUNNER_RESULT_LOG)) {
                    const item = RUNNER_RESULT_LOG[scope.iteration_id];

                    if (item) {
                        item.visualizer_html = { status, html };
                    }
                }
            }
        }

        // 断言自定义拓展规则（100% 兼容postman）
        chai__default["default"].use(() => {
            chaiApipost__default["default"](chai__default["default"]);
        });

        // 执行脚本
        /**
             * code js 代码（当前仅支持 js 代码，后续支持多种语言，如 python 等）
             * scope 当前变量环境对象，里面包含
             * {
            //  *      variables   临时变量 此数据为临时数据，无需存库
             *      globals     公共变量
             *      environment     环境变量
             *      collectionVariables     目录变量
            //  *      iterationData   流程测试时的数据变量 此数据为临时数据，无需存库
             *      request     请求参数
             *      response    响应参数
             *      cookie      cookie
             *      info        请求运行相关信息
             * }
             *
             * callback 回调
         * */
        async function execute(code, scope, eventName, callback) {
            scope = lodash__default["default"].isPlainObject(scope) ? lodash__default["default"].cloneDeep(scope) : {};

            // 初始化数据库中的当前变量值 init
            if (typeof aptScripts === 'object') {
                Object.keys(variablesScope).forEach((key) => {
                    if (lodash__default["default"].isObject(aptScripts[key]) && lodash__default["default"].isFunction(aptScripts[key].toObject) && ['iterationData', 'variables'].indexOf(key) > -1) {
                        lodash__default["default"].assign(variablesScope[key], aptScripts[key].toObject());
                    }
                });
            }

            // pm 对象
            const pm = {};

            // info, 请求、响应、cookie, iterationData
            new Array('info', 'request', 'response', 'cookie', 'iterationData').forEach((key) => {
                if (lodash__default["default"].indexOf(['request', 'response'], key) > -1) {
                    switch (key) {
                        case 'request':
                            // console.log(scope.script_request);
                            if (lodash__default["default"].has(scope, 'script_request') && lodash__default["default"].isObject(scope.script_request)) {
                                Object.defineProperty(scope.script_request, 'to', {
                                    get() {
                                        return chai__default["default"].expect(this).to;
                                    },
                                });

                                Object.defineProperty(pm, key, {
                                    configurable: true,
                                    value: scope.script_request,
                                });
                            }
                            break;
                        case 'response':
                            if (lodash__default["default"].has(scope, `response.data.${key}`) && lodash__default["default"].isObject(scope.response.data[key])) {
                                if (scope.response.data[key].hasOwnProperty('rawBody')) {
                                    let json = {};

                                    try {
                                        json = json5__default["default"].parse(scope.response.data[key].rawBody);
                                    } catch (e) { }

                                    Object.defineProperty(scope.response.data[key], 'json', {
                                        configurable: true,
                                        value() {
                                            return lodash__default["default"].cloneDeep(json);
                                        },
                                    });

                                    Object.defineProperty(scope.response.data[key], 'text', {
                                        configurable: true,
                                        value() {
                                            return scope.response.data[key].rawBody;
                                        },
                                    });
                                }

                                Object.defineProperty(scope.response.data[key], 'to', {
                                    get() {
                                        return chai__default["default"].expect(this).to;
                                    },
                                });

                                Object.defineProperty(pm, key, {
                                    configurable: true,
                                    value: scope.response.data[key],
                                });
                            }
                            break;
                    }
                } else if (lodash__default["default"].isObject(scope[key])) {
                    switch (key) {
                        case 'iterationData':
                            lodash__default["default"].assign(variablesScope.iterationData, scope[key]);
                            break;
                        case 'info':
                            lodash__default["default"].assign(scope[key], {
                                iteration: scope.iteration,
                                iterationCount: scope.iterationCount,
                                eventName,
                            });
                            break;
                    }

                    Object.defineProperty(pm, key, {
                        configurable: true,
                        value: scope[key],
                    });
                }
            });

            // 变量相关
            Object.keys(variablesScope).forEach((type) => {
                Object.defineProperty(pm, type, {
                    configurable: true,
                    value: dynamicVariables[type],
                });
            });

            if (lodash__default["default"].isObject(pm.variables)) {
                Object.defineProperty(pm.variables, 'getName', {
                    configurable: true,
                    value() {
                        return scope.env_name;
                    },
                });

                Object.defineProperty(pm.variables, 'getPreUrl', {
                    configurable: true,
                    value() {
                        const api_server_id = getCollectionServerId(scope?.target_id, scope?.collection);
                        const script_pre_env_url = scope?.env_pre_urls?.[api_server_id];
                        if (lodash__default["default"].isUndefined(script_pre_env_url)) {
                            return scope.env_pre_url;
                        }
                        return script_pre_env_url

                    },
                });

                Object.defineProperty(pm.variables, 'getCollection', {
                    configurable: true,
                    value() {
                        return scope.environment;
                    },
                });

                Object.defineProperty(pm.environment, 'getName', {
                    configurable: true,
                    value() {
                        return scope.env_name;
                    },
                });

                Object.defineProperty(pm.environment, 'getPreUrl', {
                    configurable: true,
                    value() {
                        const api_server_id = getCollectionServerId(scope?.target_id, scope?.collection);
                        const script_pre_env_url = scope?.env_pre_urls?.[api_server_id];
                        if (lodash__default["default"].isUndefined(script_pre_env_url)) {
                            return scope.env_pre_url;
                        }
                        return script_pre_env_url
                    },
                });

                Object.defineProperty(pm.environment, 'getCollection', {
                    configurable: true,
                    value() {
                        return scope.environment;
                    },
                });
            }

            // 请求参数相关
            if (typeof scope !== 'undefined' && lodash__default["default"].isObject(scope) && lodash__default["default"].has(scope, 'request.request')) {
                // 更新日志

                if (lodash__default["default"].isObject(RUNNER_RESULT_LOG)) {
                    const item = RUNNER_RESULT_LOG[scope.iteration_id];

                    if (item) {
                        Object.defineProperty(pm, 'setRequestQuery', {
                            configurable: true,
                            value(key, value) {
                                if (lodash__default["default"].trim(key) != '') {
                                    if (!lodash__default["default"].has(item, 'beforeRequest.query')) {
                                        lodash__default["default"].set(item, 'beforeRequest.query', []);
                                    }

                                    item.beforeRequest.query.push({
                                        action: 'set',
                                        key,
                                        value,
                                    });
                                }
                            },
                        });

                        Object.defineProperty(pm, 'removeRequestQuery', {
                            configurable: true,
                            value(key) {
                                if (lodash__default["default"].trim(key) != '') {
                                    if (!lodash__default["default"].has(item, 'beforeRequest.query')) {
                                        lodash__default["default"].set(item, 'beforeRequest.query', []);
                                    }

                                    item.beforeRequest.query.push({
                                        action: 'remove',
                                        key,
                                    });
                                }
                            },
                        });

                        Object.defineProperty(pm, 'setRequestHeader', {
                            configurable: true,
                            value(key, value) {
                                if (lodash__default["default"].trim(key) != '') {
                                    if (!lodash__default["default"].has(item, 'beforeRequest.header')) {
                                        lodash__default["default"].set(item, 'beforeRequest.header', []);
                                    }

                                    item.beforeRequest.header.push({ // fix bug for 7.0.8
                                        action: 'set',
                                        key: String(key),
                                        value: String(value),
                                    });
                                }
                            },
                        });

                        Object.defineProperty(pm, 'removeRequestHeader', {
                            configurable: true,
                            value(key) {
                                if (lodash__default["default"].trim(key) != '') {
                                    if (!lodash__default["default"].has(item, 'beforeRequest.header')) {
                                        lodash__default["default"].set(item, 'beforeRequest.header', []);
                                    }

                                    item.beforeRequest.header.push({
                                        action: 'remove',
                                        key,
                                    });
                                }
                            },
                        });

                        Object.defineProperty(pm, 'setRequestBody', {
                            configurable: true,
                            value(key, value) {
                                if (lodash__default["default"].trim(key) != '') {
                                    if (!lodash__default["default"].has(item, 'beforeRequest.body')) {
                                        lodash__default["default"].set(item, 'beforeRequest.body', []);
                                    }

                                    item.beforeRequest.body.push({
                                        action: 'set',
                                        key,
                                        value,
                                    });
                                }
                            },
                        });

                        Object.defineProperty(pm, 'removeRequestBody', {
                            configurable: true,
                            value(key) {
                                if (lodash__default["default"].trim(key) != '') {
                                    if (!lodash__default["default"].has(item, 'beforeRequest.body')) {
                                        lodash__default["default"].set(item, 'beforeRequest.body', []);
                                    }

                                    item.beforeRequest.body.push({
                                        action: 'remove',
                                        key,
                                    });
                                }
                            },
                        });
                    }
                }
            }

            // expert
            Object.defineProperty(pm, 'expect', {
                configurable: true,
                value: chai__default["default"].expect,
            });

            // test
            Object.defineProperty(pm, 'test', {
                configurable: true,
                value(desc, callback) {
                    try {
                        callback();
                        emitAssertResult('success', desc, '成功', scope);
                    } catch (e) {
                        emitAssertResult('error', desc, e.toString().replace('AssertionError', '断言校验失败'), scope);
                    }
                },
            });

            // assert
            Object.defineProperty(pm, 'assert', {
                configurable: true,
                value(assert) {
                    try {
                        const _response = lodash__default["default"].cloneDeep(pm.response);

                        if (lodash__default["default"].isFunction(_response.json)) {
                            _response.json = _response.json();
                        }

                        chai__default["default"].assert.isTrue(new Function('response', 'request', 'window', `return ${String(assert)}`)(_response, lodash__default["default"].cloneDeep(pm.request)));
                        emitAssertResult('success', String(assert), '成功', scope);
                        return true; // fixed bug
                    } catch (e) {
                        emitAssertResult('error', String(assert), e.toString().replace('AssertionError', '断言校验失败').replace('expected false to be true', '表达式不成立'), scope);
                        return false; // fixed bug
                    }
                },
            });

            // 发送方法
            Object.defineProperty(pm, 'sendRequest', {
                configurable: true,
                value: ajaxForNode__default["default"], // fix bug
            });

            // 可视化
            Object.defineProperty(pm, 'visualizer', {
                configurable: true,
                value: {
                    set: (template, data) => {
                        try {
                            const html = artTemplate__default["default"].render(template, data);
                            emitVisualizerHtml('success', `<link rel="stylesheet" href="https://img.cdn.apipost.cn/docs/css7/content-v7.css?20220909" type="text/css" media="screen"> ${html}`, scope);
                        } catch (e) {
                            emitVisualizerHtml('error', e.toString(), scope);
                        }
                    },
                },
            });

            // 执行外部程序
            enableUnSafeShell && Object.defineProperty(pm, 'execute', {
                configurable: true,
                value: function (file, args, extra) {
                    if (lodash__default["default"].isString(file)) {
                        try {
                            try {
                                fs__default["default"].accessSync(file);
                            } catch (e) {
                                let externalPrograms = _.get(option, 'requester.externalPrograms');

                                if (_.isString(externalPrograms)) {
                                    try {
                                        file = path__default["default"].join(path__default["default"].resolve(externalPrograms), file);
                                    } catch (e) {

                                        file = path__default["default"].join(path__default["default"].resolve(apipostInsideTools__default["default"].getCachePath()), `apipost`, 'ExternalPrograms', file);

                                    }
                                } else {
                                    file = path__default["default"].join(path__default["default"].resolve(apipostInsideTools__default["default"].getCachePath()), `apipost`, 'ExternalPrograms', file);
                                }
                            }

                            let command = ``;
                            switch (lodash__default["default"].toLower(file.substr(file.lastIndexOf(".")))) {
                                case '.go':
                                    command = `go run `;
                                    break;
                                case '.php':
                                    command = `php -f `;
                                    break;
                                case '.py':
                                    command = `python `;
                                    break;
                                case '.js':
                                    command = `node `;
                                    break;
                                case '.jar':
                                    if (lodash__default["default"].isObject(extra) && lodash__default["default"].isObject(process) && lodash__default["default"].isString(lodash__default["default"].get(process, 'resourcesPath'))) {
                                        let className = lodash__default["default"].get(extra, 'className');
                                        let method = lodash__default["default"].get(extra, 'method');

                                        if (lodash__default["default"].isString(className) && lodash__default["default"].isString(method)) {
                                            let jarPath = path__default["default"].join(path__default["default"].resolve(process.resourcesPath), `app`, `jar-main-1.0-SNAPSHOT.jar`);
                                            let para = new Buffer(JSON.stringify({ "methodName": method, "args": args })).toString('base64');
                                            command = `java -jar ${jarPath}  ${file} ${className} '${para}'`;
                                        }
                                    } else {
                                        command = `java -jar `;
                                    }

                                    break;
                                case '.sh':
                                    command = `bash `;
                                    break;
                                case '.ruby':
                                    command = `ruby `;
                                    break;
                                case '.lua':
                                    command = `lua `;
                                    break;
                                case '.bsh':
                                    command = `bsh `;
                            }

                            if (command != '') {
                                if (lodash__default["default"].toLower(file.substr(file.lastIndexOf("."))) == '.jar' && lodash__default["default"].isObject(extra) && lodash__default["default"].isObject(process) && lodash__default["default"].isString(lodash__default["default"].get(process, 'resourcesPath'))) {
                                    return String(child_process__default["default"].execSync(`${command}`))
                                } else {
                                    return String(child_process__default["default"].execSync(`${command} ${file} ${lodash__default["default"].join(args, ' ')}`))
                                }
                            }
                        }
                        catch (e) { return String(e.stderr) }
                    }
                },
            });

            Object.defineProperty(pm, 'Visualizing', {
                configurable: true,
                value: (template, data) => {
                    try {
                        const html = artTemplate__default["default"].render(template, data);
                        // console.log(html, template);
                        emitVisualizerHtml('success', `<link rel="stylesheet" href="https://img.cdn.apipost.cn/docs/css7/content-v7.css?20220909" type="text/css" media="screen"> ${html}`, scope);
                    } catch (e) {
                        // console.log(e);
                        emitVisualizerHtml('error', e.toString(), scope);
                    }
                },
            });

            Object.defineProperty(pm, 'getData', { // 此方法为兼容 postman ，由于流程差异，暂时不支持
                configurable: true,
                value(callback) {
                    // @todo
                },
            });

            // 跳过下面的流程直接到执行指定接口
            Object.defineProperty(pm, 'setNextRequest', { // 此方法为兼容 postman ，由于流程差异，暂时不支持
                configurable: true,
                value(target_id) {
                    // @todo
                },
            });

            // 执行
            try {
                // const $ = {};

                jquery__default["default"].md5 = function (str) { // 兼容旧版
                    return cryptoJs__default["default"].MD5(str).toString();
                };

                jquery__default["default"].ajax = await ajaxForNode__default["default"];

                // fix bug
                code = `(async function () {
          ${code}
        })()`;
                let scriptTimeout = _.toNumber(_.get(option, 'requester.timeoutScript'));

                if (scriptTimeout <= 0) {
                    scriptTimeout = 5000;
                }
                await (new vm2__default["default"].VM({
                    timeout: scriptTimeout,
                    sandbox: lodash__default["default"].assign({
                        ...{ nodeAjax: ajaxForNode__default["default"] },
                        ...{ pm },
                        ...{ chai: chai__default["default"] },
                        ...{ emitAssertResult },
                        ...{ JSON5: json5__default["default"] },
                        ...{ _: lodash__default["default"] },
                        ...{ Mock: apipostMock__default["default"] },
                        ...{ uuid: uuid__default["default"] },
                        ...{ jsonpath: jsonpath__default["default"] },
                        ...{ CryptoJS: cryptoJs__default["default"] },
                        // ...{ $ },
                        ...{ x2js: x2js__default["default"] },
                        JSEncrypt: jsencryptNode__default["default"],
                        ...{ moment: moment__default["default"] },
                        ...{ dayjs: dayjs__default["default"] },
                        JSON, // 增加 JSON 方法 // fixed JSON5 bug
                        console: consoleFn,
                        print: consoleFn.log,
                        async: async__default["default"],
                        FormData: formData__default["default"],
                        require: function () { },
                        sm2, // fix bug for 7.0.8
                        sm3, // fix bug for 7.0.8
                        sm4, // fix bug for 7.0.8
                        csv2array: testdataToApipostJson__default["default"],
                        mysql: mysql2__default["default"],
                        mssql: mssql__default["default"], ClickHouse, pgClient,
                        fs: enableUnSafeShell ? fs__default["default"] : {},
                        path: path__default["default"], json2csv: json2Csv__default["default"], // for 7.0.13
                        xml2json(xml) {
                            return (new x2js__default["default"]()).xml2js(xml);
                        },
                        uuidv4() {
                            return uuid__default["default"].v4();
                        },
                        URL: url__default["default"],
                        ...{ uuid: uuid__default["default"] },
                        ...{ aTools: apipostTools__default["default"] },
                        ...{ validCookie: checkValidCookie__default["default"] },
                        ...{ urlJoin: urlJoin__default["default"] },
                        urljoins, // fix bug for 7.0.8
                        apt: pm,
                        fox: pm,
                        $: jquery__default["default"],
                        // Promise,
                        request: pm.request ? lodash__default["default"].cloneDeep(pm.request) : {},
                        response: pm.response ? lodash__default["default"].assign(pm.response, { json: lodash__default["default"].isFunction(pm.response.json) ? pm.response.json() : pm.response.json }) : {},
                        expect: chai__default["default"].expect,
                        sleep: atomicSleep__default["default"],
                        // child_process
                        // sleep(ms) {
                        //   const end = Date.now() + parseInt(ms);
                        //   while (true) {
                        //     if (Date.now() > end) {
                        //       return;
                        //     }
                        //   }
                        // },
                    }, variablesScope),
                })).run(new vm2__default["default"].VMScript(code));
                typeof callback === 'function' && callback(null, pm.response);
            } catch (err) {
                emitTargetPara({
                    action: 'SCRIPT_ERROR',
                    eventName,
                    data: `${eventName == 'pre_script' ? '预执行' : '后执行'}脚本语法错误: ${err.toString()}`,
                }, scope);
                typeof callback === 'function' && callback(err.toString());
            }
        }

        lodash__default["default"].assign(this, {
            ...{ execute },
            ...{ getAllInsideVariables },
            ...{ getAllDynamicVariables },
            ...{ dynamicVariables },
            ...{ variablesScope },
            ...{ replaceIn },
        });
    };

    const mySandbox = new Sandbox();

    // sleep 延迟方法
    function sleepDelay(ms) {
        atomicSleep__default["default"](ms);
    }

    // 根据测试条件返回布尔值
    function returnBoolean(exp, compare, value) {
        let bool = false;
        if (exp === '') { // fix bug
            return compare == 'null';
        }

        // get request
        if (typeof exp == 'string' && lodash__default["default"].has(PREV_REQUEST, 'request') && lodash__default["default"].startsWith(exp, `{{request\.`) && lodash__default["default"].endsWith(exp, '}}')) {
            let _path = String(lodash__default["default"].trimEnd(lodash__default["default"].trimStart(exp, `{{request`), '}}'));

            if (_path.substr(0, 1) == '.') {
                _path = _path.substr(1);
            }

            exp = lodash__default["default"].get(PREV_REQUEST.request, _path);
        }

        if (typeof value == 'string' && lodash__default["default"].has(PREV_REQUEST, 'request') && lodash__default["default"].startsWith(value, `{{request\.`) && lodash__default["default"].endsWith(value, '}}')) {
            let _path = String(lodash__default["default"].trimEnd(lodash__default["default"].trimStart(value, `{{request`), '}}'));

            if (_path.substr(0, 1) == '.') {
                _path = _path.substr(1);
            }

            value = lodash__default["default"].get(PREV_REQUEST.request, _path);
        }

        // get response
        if (typeof exp == 'string' && lodash__default["default"].has(PREV_REQUEST, 'response') && lodash__default["default"].startsWith(exp, `{{response\.`) && lodash__default["default"].endsWith(exp, '}}')) {
            let _path = String(lodash__default["default"].trimEnd(lodash__default["default"].trimStart(exp, `{{response`), '}}'));

            if (_path.substr(0, 1) == '.') {
                _path = _path.substr(1);
            }

            exp = lodash__default["default"].get(PREV_REQUEST.response, _path);
        }

        if (typeof value == 'string' && lodash__default["default"].has(PREV_REQUEST, 'response') && lodash__default["default"].startsWith(value, `{{response\.`) && lodash__default["default"].endsWith(value, '}}')) {
            let _path = String(lodash__default["default"].trimEnd(lodash__default["default"].trimStart(value, `{{response`), '}}'));

            if (_path.substr(0, 1) == '.') {
                _path = _path.substr(1);
            }

            value = lodash__default["default"].get(PREV_REQUEST.response, _path);
        }

        switch (compare) {
            case 'eq':
                bool = exp == value;
                break;
            case 'uneq':
                bool = exp != value;
                break;
            case 'gt':
                bool = lodash__default["default"].gt(Number(exp), Number(value));
                break;
            case 'gte':
                bool = lodash__default["default"].gte(Number(exp), Number(value));
                break;
            case 'lt':
                bool = lodash__default["default"].lt(Number(exp), Number(value));
                break;
            case 'lte':
                bool = lodash__default["default"].lte(Number(exp), Number(value));
                break;
            case 'includes':
                bool = lodash__default["default"].includes(exp, value) || lodash__default["default"].includes(exp, Number(value)); // fix bug
                break;
            case 'unincludes':
                bool = !lodash__default["default"].includes(exp, value) && !lodash__default["default"].includes(exp, Number(value)); // fix bug
                break;
            case 'null':
                bool = lodash__default["default"].isNull(exp, value);
                break;
            case 'notnull':
                bool = !lodash__default["default"].isNull(exp, value);
                break;
        }

        return bool;
    }

    // 获取某接口的 所有父target
    function getParentTargetIDs(collection, target_id, parent_ids = []) {
        if (lodash__default["default"].isArray(collection)) {
            const item = lodash__default["default"].find(collection, lodash__default["default"].matchesProperty('target_id', target_id));

            if (item) {
                parent_ids.push(item.parent_id);
                getParentTargetIDs(collection, item.parent_id, parent_ids);
            }
        }

        return parent_ids;
    }

    // 获取 指定 event_id 的 initDefinitions 的所有父亲ID
    function getInitDefinitionsParentIDs(event_id, initDefinitions = []) {
        const definitionArr = [];

        (function convertArray(initDefinitions) {
            if (lodash__default["default"].isArray(initDefinitions)) {
                initDefinitions.forEach((item) => {
                    definitionArr.push({
                        event_id: item.event_id,
                        parent_id: item.parent_id,
                    });

                    if (lodash__default["default"].isArray(item.children)) {
                        convertArray(item.children);
                    }
                });
            }
        }(initDefinitions));

        const parentArr = [];

        (function getParentArr(event_id) {
            definitionArr.forEach((item) => {
                if (item.event_id == event_id) {
                    // if (uuid.validate(item.parent_id)) {
                    if (item.parent_id != '0') {
                        parentArr.push(item.parent_id);
                        getParentArr(item.parent_id);
                    }
                }
            });
        }(event_id));

        return parentArr;
    }

    // 获取某接口的详细信息
    function getItemFromCollection(collection, target_id) {
        return lodash__default["default"].find(collection, lodash__default["default"].matchesProperty('target_id', target_id));
    }

    // 计算runtime 结果
    function calculateRuntimeReport(log, initDefinitions = [], report_id = '', option = {}) {
        log = Object.values(log);

        // 说明： 本api统计数字均已去重
        // 接口去重后的api集合
        const _uniqLog = lodash__default["default"].uniqWith(log, (source, dist) => lodash__default["default"].isEqual(source.target_id, dist.target_id));

        // 接口未去重后的忽略集合
        const _ignoreLog = lodash__default["default"].filter(log, item => item.http_error == -2);

        // 接口去重后的忽略集合
        const _uniqIgnoreLog = lodash__default["default"].uniqWith(_ignoreLog, (source, dist) => lodash__default["default"].isEqual(source.target_id, dist.target_id));

        // 接口未去重后的http失败集合
        const _httpErrorLog = lodash__default["default"].filter(log, item => item.http_error == 1);

        // 接口去重后的http失败集合
        const _uniqHttpErrorLog = lodash__default["default"].uniqWith(_httpErrorLog, (source, dist) => lodash__default["default"].isEqual(source.target_id, dist.target_id));

        // 接口未去重后的assert失败集合
        const _assertErrorLog = lodash__default["default"].filter(log, item => lodash__default["default"].find(item.assert, lodash__default["default"].matchesProperty('status', 'error')));

        // 接口去重后的assert失败集合
        const _uniqAssertErrorLog = lodash__default["default"].uniqWith(_assertErrorLog, (source, dist) => lodash__default["default"].isEqual(source.target_id, dist.target_id));

        // 接口未去重后的assert成功集合
        const _assertPassedLog = lodash__default["default"].filter(log, item => lodash__default["default"].size(item.assert) > 0 && !lodash__default["default"].find(item.assert, lodash__default["default"].matchesProperty('status', 'error')));

        // 接口去重后的assert成功集合
        const _uniqAssertPassedLog = lodash__default["default"].uniqWith(_assertPassedLog, (source, dist) => lodash__default["default"].isEqual(source.target_id, dist.target_id));

        // 接口未去重后的http成功集合
        const _httpPassedLog = lodash__default["default"].filter(log, item => item.http_error == -1 && !lodash__default["default"].find(_uniqHttpErrorLog, lodash__default["default"].matchesProperty('target_id', item.target_id)));

        // 接口去重后的http成功集合
        const _uniqHttpPassedLog = lodash__default["default"].uniqWith(_httpPassedLog, (source, dist) => lodash__default["default"].isEqual(source.target_id, dist.target_id));
        // console.log(_uniqHttpPassedLog);
        // 计算 总api数
        const totalCount = lodash__default["default"].size(_uniqLog);

        // 计算 未忽略的总api数
        const totalEffectiveCount = lodash__default["default"].subtract(totalCount, lodash__default["default"].size(_uniqIgnoreLog));

        // 计算 http 错误个数
        const httpErrorCount = lodash__default["default"].size(_uniqHttpErrorLog);

        // 计算 http 成功个数
        const httpPassedCount = lodash__default["default"].size(_uniqHttpPassedLog);

        // 计算 assert 错误个数
        const assertErrorCount = lodash__default["default"].size(_uniqAssertErrorLog);

        // 计算 assert 错误个数
        const assertPassedCount = lodash__default["default"].size(_uniqAssertPassedLog);

        // 计算 忽略接口 个数
        const ignoreCount = lodash__default["default"].size(_uniqIgnoreLog);

        // 获取 event 事件状态
        const eventResultStatus = {};

        Object.values(log).forEach((item) => {
            // 计算各个event的状态 [ignore, failure, passed]
            if (lodash__default["default"].isArray(initDefinitions)) {
                const parent_ids = getInitDefinitionsParentIDs(item.event_id, initDefinitions);

                if (lodash__default["default"].find(item.assert, lodash__default["default"].matchesProperty('status', 'error'))) {
                    item.assert_error = 1;
                } else {
                    item.assert_error = -1;
                }

                if (item.http_error == 1 || lodash__default["default"].find(item.assert, lodash__default["default"].matchesProperty('status', 'error'))) { // failure
                    eventResultStatus[item.event_id] = 'failure';
                    parent_ids.forEach((parent_id) => {
                        if (lodash__default["default"].indexOf(Object.keys(eventResultStatus), parent_id) == -1) {
                            eventResultStatus[parent_id] = 'failure';
                        }
                    });
                } else if (item.http_error == -2) {
                    eventResultStatus[item.event_id] = 'ignore';
                    parent_ids.forEach((parent_id) => {
                        if (lodash__default["default"].indexOf(Object.keys(eventResultStatus), parent_id) == -1) {
                            eventResultStatus[parent_id] = 'ignore';
                        }
                    });
                } else if (item.http_error == -1) {
                    eventResultStatus[item.event_id] = 'passed';
                    parent_ids.forEach((parent_id) => {
                        if (lodash__default["default"].indexOf(Object.keys(eventResultStatus), parent_id) == -1) {
                            eventResultStatus[parent_id] = 'passed';
                        }
                    });
                }
            }
        });

        const definitionList = [];

        (function convertInitDefinitions(initDefinitions) {
            initDefinitions.forEach((item) => {
                if (lodash__default["default"].isString(item.test_id)) {
                    definitionList.push({
                        event_id: item.event_id,
                        parent_event_id: item.parent_id,
                        data: (item.type == 'api' || item.type == 'sample') ? item.request : item.condition,
                        enabled: item.enabled,
                        project_id: item.project_id,
                        sort: item.sort,
                        test_id: item.test_id,
                        type: item.type,
                        report_id,
                        runtime: item.runtime,
                        runtime_status: eventResultStatus[item.event_id],
                    });
                }

                if (lodash__default["default"].isArray(item.children)) {
                    convertInitDefinitions(item.children);
                }
            });
        }(initDefinitions));

        // 计算 received_data， total_response_time
        let _total_received_data = 0,
            _total_response_time = 0,
            _total_response_count = 0;

        lodash__default["default"].forEach(log, (item) => {
            if (lodash__default["default"].has(item, 'response.data.response.responseSize')) {
                _total_response_count++;
                _total_received_data = lodash__default["default"].add(_total_received_data, Number(item.response.data.response.responseSize));
                _total_response_time = lodash__default["default"].add(_total_response_time, Number(item.response.data.response.responseTime));
            }
        });

        if (typeof option.env === 'undefined') {
            option.env = {
                env_id: option.env_id ? option.env_id : -1,
                env_name: option.env_name,
                env_pre_url: option.env_pre_url,
                env_pre_urls: option?.env_pre_urls,
            };
        } else {
            option.env_id = option.env.env_id;
            option.env_name = option.env.env_name;
            option.env_pre_url = option.env.env_pre_url;
            option.env_pre_urls = option.env?.env_pre_urls;
        }

        const report = {
            combined_id: option.combined_id,
            report_id,
            report_name: option.default_report_name,
            env_id: option.env.env_id,
            env_name: option.env.env_name,
            env_pre_url: option.env.env_pre_url,
            env_pre_urls: option.env?.env_pre_urls,
            user: option.user,
            total_count: totalCount,
            total_effective_count: totalEffectiveCount,
            ignore_count: ignoreCount,
            total_received_data: lodash__default["default"].floor(_total_received_data, 2),
            total_response_time: lodash__default["default"].floor(_total_response_time, 2),
            average_response_time: lodash__default["default"].floor(lodash__default["default"].divide(_total_response_time, _total_response_count), 2),
            http_errors: _httpErrorLog,
            assert_errors: _assertErrorLog,
            ignore_errors: _ignoreLog,
            http: {
                passed: httpPassedCount,
                passed_per: lodash__default["default"].floor(lodash__default["default"].divide(httpPassedCount, totalCount), 2),
                failure: httpErrorCount,
                failure_per: lodash__default["default"].floor(lodash__default["default"].divide(httpErrorCount, totalCount), 2),
            },
            assert: {
                passed: assertPassedCount,
                passed_per: lodash__default["default"].floor(lodash__default["default"].divide(assertPassedCount, totalCount), 2),
                failure: assertErrorCount,
                failure_per: lodash__default["default"].floor(lodash__default["default"].divide(assertErrorCount, totalCount), 2),
            },
            start_time: startTime,
            end_time: dayjs__default["default"]().format('YYYY-MM-DD HH:mm:ss'),
            long_time: `${lodash__default["default"].floor(lodash__default["default"].divide(Date.now() - startTimeStamp, 1000), 2)} 秒`,
            children: [],
        };

        if (!lodash__default["default"].has(report, 'user.nick_name')) {
            lodash__default["default"].set(report, 'user.nick_name', '匿名');
        }
        if (uuid__default["default"].validate(option.combined_id) && lodash__default["default"].isArray(option.test_events)) { // 测试套件
            lodash__default["default"].assign(report, {
                type: 'combined',
                test_id: lodash__default["default"].isArray(option.test_events) ? (lodash__default["default"].map(option.test_events, o => o.test_id)) : [option.test_events.test_id],
            });

            option.test_events.forEach((test_event) => {
                report.children.push(calculateRuntimeReport(lodash__default["default"].filter(log, o => o.test_id == test_event.test_id), initDefinitions, report_id, lodash__default["default"].assign(option, {
                    combined_id: 0,
                    test_events: test_event,
                    default_report_name: test_event.name,
                })));
            });
        } else { // 单测试用例
            const _test_id = lodash__default["default"].isArray(option.test_events) ? option.test_events[0].test_id : option.test_events.test_id;
            lodash__default["default"].assign(report, {
                type: 'single',
                test_id: _test_id,
                event_status: eventResultStatus,
                test_events: lodash__default["default"].filter(definitionList, o => o.test_id == _test_id),
            });
        }
        log = null;
        return report;
    }


    // 参数初始化
    function runInit() {
        RUNNER_ERROR_COUNT = 0;
        // RUNNER_TOTAL_COUNT = 0
        startTime = dayjs__default["default"]().format('YYYY-MM-DD HH:mm:ss'); // 开始时间
        startTimeStamp = Date.now(); // 开始时间戳
        RUNNER_RESULT_LOG = {};
    }

    // 停止 run
    function stop(report_id, message) {
        RUNNER_STOP[report_id] = 1;

        emitRuntimeEvent({
            action: 'stop',
            report_id,
            message,
        });
    }

    let startTime = dayjs__default["default"]().format('YYYY-MM-DD HH:mm:ss'), // 开始时间
        startTimeStamp = Date.now(), // 开始时间戳
        initDefinitions = [], // 原始colletion
        RUNNER_RUNTIME_POINTER = 0;

    // start run
    const PREV_REQUEST = {
        request: {},
        response: {},
        iterationData: {},
        iterationCount: 0
    };

    async function run(definitions, option = {}, initFlag = 0, loopCount = 0) {
        // console.log(option, 'option1')
        option = lodash__default["default"].assign({
            project: {},
            collection: [], // 当前项目的所有接口列表
            environment: {}, // 当前环境变量
            globals: {}, // 当前公共变量
            iterationData: [], // 当前迭代的excel导入数据
            iterationCount: loopCount || 1, // 当前迭代次数
            ignoreError: 1, // 遇到错误忽略
            sleep: 0, // 每个任务的间隔时间
            requester: {}, // 发送模块的 options
        }, option);
        // console.log(option, 'option2')
        let { RUNNER_REPORT_ID, scene, project, cookies, collection, iterationData, combined_id, test_events, default_report_name, user, env, env_name, env_pre_url, env_pre_urls, environment, globals, iterationCount, ignoreError, ignore_error, enable_sandbox, sleep, requester } = option;

        if (typeof ignoreError === 'undefined') {
            ignoreError = ignore_error ? !!ignore_error : 0;
        } else {
            ignoreError = !!ignoreError;
        }
        ignore_error = ignoreError ? 1 : 0;
        enable_sandbox = typeof enable_sandbox === 'undefined' ? -1 : enable_sandbox; // fix bug


        if (typeof env === 'undefined') {
            env = {
                env_name,
                env_pre_url,
                env_pre_urls
            };
        } else {
            env_name = env.env_name;
            env_pre_url = env.env_pre_url;
            env_pre_urls = env.env_pre_urls;
        }

        if (initFlag == 0) { // 初始化参数
            if (lodash__default["default"].size(RUNNER_RESULT_LOG) > 0) { // 当前有任务时，拒绝新任务
                return;
            }

            // 设置sandbox的 environment变量 和 globals 变量
            // fix bug for 7.0.8
            new Array('environment', 'globals').forEach((func) => {
                if (lodash__default["default"].isObject(option[func]) && lodash__default["default"].isObject(mySandbox.dynamicVariables[func]) && lodash__default["default"].isFunction(mySandbox.dynamicVariables[func].set)) {
                    for (const [key, value] of Object.entries(option[func])) {
                        mySandbox.dynamicVariables[func].set(key, value, false);
                    }
                }
            });

            runInit();
            RUNNER_STOP[RUNNER_REPORT_ID] = 0;
            RUNNER_TOTAL_COUNT = typeof definitions[0] === 'object' ? definitions[0].RUNNER_TOTAL_COUNT : 0;
            RUNNER_RUNTIME_POINTER = 0;
            initDefinitions = definitions;

            if (RUNNER_TOTAL_COUNT <= 0) {
                return;
            }
        } else if (RUNNER_STOP[RUNNER_REPORT_ID] > 0) {
            RUNNER_RESULT_LOG = definitions = option = collection = initDefinitions = null;
            return;
        }

        // 使用场景 auto_test/request
        if (lodash__default["default"].isUndefined(scene)) {
            scene = 'auto_test';
        }

        // 兼容 单接口请求 和 自动化测试
        if (!uuid__default["default"].validate(combined_id)) {
            combined_id = '0';
        }

        if (!lodash__default["default"].isObject(test_events)) {
            test_events = {
                test_id: uuid__default["default"].v4(),
                name: '未命名',
            };
        }

        if (!lodash__default["default"].isString(default_report_name)) {
            default_report_name = '默认自动化测试报告';
        }

        if (!lodash__default["default"].isObject(user)) {
            user = {
                uuid: '-1',
                nick_name: '匿名',
            };
        }

        if (!lodash__default["default"].isArray(iterationData)) {  // fixed iterationData 兼容
            if (lodash__default["default"].isObject(iterationData)) {
                const _interData = lodash__default["default"].values(iterationData);
                if (lodash__default["default"].isArray(_interData) && lodash__default["default"].isArray(_interData[0])) {
                    iterationData = _interData[0];
                } else {
                    iterationData = [];
                }
            } else {
                iterationData = [];
            }
        }

        if (typeof iterationCount === 'undefined') {
            iterationCount = loopCount || 1;
        }

        // 自动替换 Mock
        const AUTO_CONVERT_FIELD_2_MOCK = typeof requester === 'object' && requester.AUTO_CONVERT_FIELD_2_MOCK > 0;


        // // 设置sandbox的 environment变量 和 globals 变量
        // new Array('environment', 'globals').forEach((func) => {
        //   console.log(option[func])
        //   if (_.isObject(option[func]) && _.isObject(mySandbox.dynamicVariables[func]) && _.isFunction(mySandbox.dynamicVariables[func].set)) {
        //     for (const [key, value] of Object.entries(option[func])) {
        //       mySandbox.dynamicVariables[func].set(key, value, false);
        //     }
        //   }
        // });

        // 发送对象
        const request = new apipostSend__default["default"](lodash__default["default"].isObject(requester) ? requester : {});

        // fix bug for 7.0.8
        if (option.sleep > 0) {
            sleepDelay(option.sleep);
        }

        // 全局断言
        const _global_asserts = lodash__default["default"].find(definitions, lodash__default["default"].matchesProperty('type', 'assert'));
        let _global_asserts_script = '';

        if (_global_asserts && lodash__default["default"].has(_global_asserts, 'data.content') && _global_asserts?.enabled > 0) {
            _global_asserts_script = _global_asserts.data.content;
        }


        if (lodash__default["default"].isArray(definitions) && definitions.length > 0) {
            for (let i = 0; i < definitions.length; i++) {
                const definition = definitions[i];
                // RUNNER_RUNTIME_POINTER

                lodash__default["default"].assign(definition, {
                    iteration_id: uuid__default["default"].v4(),  // 每次执行单任务的ID
                    iteration: loopCount,
                    iterationData: iterationData[loopCount] ? iterationData[loopCount] : iterationData[0],
                    ...{ iterationCount },
                    ...{ env_name },
                    ...{ env_pre_url },
                    ...{ env_pre_urls },
                    ...{ collection },
                    ...{ environment },
                    ...{ globals },
                });

                lodash__default["default"].set(PREV_REQUEST, 'iterationCount', loopCount);

                if (lodash__default["default"].isObject(definition.iterationData)) {
                    lodash__default["default"].set(PREV_REQUEST, 'iterationData', definition.iterationData);
                    for (const [key, value] of Object.entries(definition.iterationData)) {
                        mySandbox.dynamicVariables.iterationData.set(key, value, false);
                    }
                } else {
                    mySandbox.dynamicVariables.iterationData.clear();
                }

                if (definition.enabled > 0) {
                    // 设置沙盒的迭代变量
                    switch (definition.type) {
                        case 'wait':
                            if (definition.condition.sleep > 0) {
                                sleepDelay(parseInt(definition.condition.sleep));
                            }
                            break;
                        case 'script':
                            // case 'assert':
                            if (lodash__default["default"].has(definition, 'data.content') && lodash__default["default"].isString(definition.data.content)) {
                                await mySandbox.execute(definition.data.content, definition, 'test', (err, res) => {
                                    if (err && ignoreError < 1) {
                                        stop(RUNNER_REPORT_ID, String(err));
                                    }
                                });
                            }
                            break;
                        case 'if':
                            lodash__default["default"].set(definition, 'runtime.condition', `${mySandbox.replaceIn(definition.condition.var)} ${definition.condition.compare} ${mySandbox.replaceIn(definition.condition.value)}`);

                            if (returnBoolean(mySandbox.replaceIn(definition.condition.var), definition.condition.compare, mySandbox.replaceIn(definition.condition.value))) {
                                await run(definition.children, option, initFlag + 1, loopCount);
                            }
                            break;
                        case 'request':
                        case 'sample':
                        case 'api':
                            if (lodash__default["default"].has(definition, 'request') && lodash__default["default"].isObject(definition.request)) {
                                // 多环境
                                const api_server_id = getCollectionServerId(definition.request.target_id, collection);

                                let temp_env = definition?.temp_env || {};
                                if (lodash__default["default"].isObject(definition.temp_env) && lodash__default["default"].isString(temp_env?.pre_url)) {

                                    env_pre_url = lodash__default["default"].trim(temp_env.pre_url);
                                } else {
                                    // 还原前置url 优先从env_pre_urls中取，取不到则从原env_pre_url中取
                                    env_pre_url = env?.env_pre_urls?.[api_server_id];
                                    if (lodash__default["default"].isUndefined(env_pre_url)) {
                                        env_pre_url = env.env_pre_url;
                                    }
                                }
                                let res = {};
                                // 拼接全局参数、目录参数、以及脚本
                                let _requestPara = {};
                                let _parent_ids = lodash__default["default"].reverse(getParentTargetIDs(collection, definition.request.target_id));
                                let _requestBody = getItemFromCollection(collection, definition.request.target_id);

                                if (_requestBody && lodash__default["default"].isObject(_requestBody)) {
                                    lodash__default["default"].assign(definition.request, {
                                        url: definition.request.url ? definition.request.url : _requestBody.request.url,
                                        request: lodash__default["default"].cloneDeep(_requestBody.request), // fix bug
                                    });

                                    _requestBody = null;
                                }

                                new Array('header', 'body', 'query', 'auth', 'pre_script', 'test', 'resful').forEach((_type) => {
                                    // 参数
                                    if (lodash__default["default"].indexOf(['header', 'body', 'query', 'resful'], _type) > -1) {
                                        if (typeof _requestPara[_type] === 'undefined') {
                                            _requestPara[_type] = _type == 'header' ? {} : [];
                                        }

                                        // 全局参数
                                        if (typeof project.request === 'object' && lodash__default["default"].isArray(project.request[_type])) {
                                            project.request[_type].forEach((item) => {
                                                if (item.is_checked > 0 && lodash__default["default"].trim(item.key) != '') {
                                                    if (_type == 'header') {
                                                        _requestPara[_type][lodash__default["default"].trim(item.key)] = item;
                                                    } else {
                                                        _requestPara[_type].push(item);
                                                    }
                                                }
                                            });
                                        }

                                        // 目录参数
                                        if (lodash__default["default"].isArray(_parent_ids) && _parent_ids.length > 0) {
                                            _parent_ids.forEach((parent_id) => {
                                                const _folder = getItemFromCollection(collection, parent_id);

                                                if (lodash__default["default"].has(_folder, 'request') && lodash__default["default"].isArray(_folder.request[_type])) {
                                                    _folder.request[_type].forEach((item) => {
                                                        if (item.is_checked > 0 && lodash__default["default"].trim(item.key) != '') {
                                                            if (_type == 'header') {
                                                                _requestPara[_type][lodash__default["default"].trim(item.key)] = item;
                                                            } else {
                                                                _requestPara[_type].push(item);
                                                            }
                                                        }
                                                    });
                                                }
                                            });
                                        }

                                        // 接口参数
                                        if (lodash__default["default"].has(definition, `request.request.${_type}.parameter`) && lodash__default["default"].isArray(definition.request.request[_type].parameter)) {
                                            definition.request.request[_type].parameter.forEach((item) => {
                                                if (item.is_checked > 0 && lodash__default["default"].trim(item.key) != '') {
                                                    if (_type == 'header') {
                                                        _requestPara[_type][lodash__default["default"].trim(item.key)] = item;
                                                    } else {
                                                        _requestPara[_type].push(item);
                                                    }
                                                }
                                            });
                                        }
                                    }

                                    // 认证
                                    if (lodash__default["default"].indexOf(['auth'], _type) > -1) {
                                        if (typeof _requestPara[_type] === 'undefined') {
                                            _requestPara[_type] = {};
                                        }

                                        // 全局认证
                                        if (lodash__default["default"].has(project, `request.['${_type}']`) && lodash__default["default"].isObject(project.request[_type]) && project.request[_type].type != 'noauth') {
                                            lodash__default["default"].assign(_requestPara[_type], project.request[_type]);
                                        }

                                        // 目录认证
                                        if (lodash__default["default"].isArray(_parent_ids) && _parent_ids.length > 0) {
                                            _parent_ids.forEach((parent_id) => {
                                                const _folder = getItemFromCollection(collection, parent_id);
                                                // console.log(_folder);
                                                if (lodash__default["default"].has(_folder, `request.['${_type}']`) && lodash__default["default"].isObject(_folder.request[_type]) && _folder.request[_type].type != 'noauth') {
                                                    // console.log(_folder.request[_type]);
                                                    lodash__default["default"].assign(_requestPara[_type], _folder.request[_type]);
                                                }
                                            });
                                        }

                                        // 接口认证
                                        if (lodash__default["default"].has(definition, `request.request.${_type}`) && lodash__default["default"].isObject(definition.request.request[_type]) && definition.request.request[_type].type != 'noauth') {
                                            lodash__default["default"].assign(_requestPara[_type], definition.request.request[_type]);
                                        }
                                    }
                                    // console.log(_requestPara);
                                    // mySandbox.replaceIn(item.key, null, AUTO_CONVERT_FIELD_2_MOCK)

                                    // 脚本
                                    if (lodash__default["default"].indexOf(['pre_script', 'test'], _type) > -1) {
                                        if (typeof _requestPara[_type] === 'undefined') {
                                            _requestPara[_type] = '';
                                        }

                                        // 全局脚本， 已兼容旧版本
                                        if (lodash__default["default"].has(project, `script.['${_type}']`) && lodash__default["default"].isString(project.script[_type]) && project.script[`${_type}_switch`] > 0) {
                                            _requestPara[_type] = `${_requestPara[_type]}\r\n${project.script[_type]}`;
                                        } else if (lodash__default["default"].has(project, `request.script.['${_type}']`) && lodash__default["default"].isString(project.request.script[_type]) && project.request.script[`${_type}_switch`] > 0) {
                                            _requestPara[_type] = `${_requestPara[_type]}\r\n${project.request.script[_type]}`;
                                        }

                                        // 目录脚本
                                        if (lodash__default["default"].isArray(_parent_ids) && _parent_ids.length > 0) {
                                            _parent_ids.forEach((parent_id) => {
                                                const _folder = getItemFromCollection(collection, parent_id);

                                                if (lodash__default["default"].has(_folder, `script.['${_type}']`) && lodash__default["default"].isString(_folder.script[_type]) && _folder.script[`${_type}_switch`] > 0) {
                                                    _requestPara[_type] = `${_requestPara[_type]}\r\n${_folder.script[_type]}`;
                                                }
                                            });
                                        }

                                        // 接口脚本
                                        if (lodash__default["default"].has(definition, `request.request.event.${_type}`) && lodash__default["default"].isString(definition.request.request.event[_type])) {
                                            _requestPara[_type] = `${_requestPara[_type]}\r\n${definition.request.request.event[_type]}`;
                                        }
                                    }
                                });

                                let _timeout = 0;
                                if (lodash__default["default"].has(option, 'requester.timeout') && lodash__default["default"].isNumber(option.requester.timeout) && option.requester.timeout >= 0) {
                                    _timeout = option.requester.timeout;
                                }

                                // script_mode
                                let _script_mode = 'none';

                                if (lodash__default["default"].has(definition, 'request.request.body.mode')) {
                                    _script_mode = definition.request.request.body.mode;
                                }

                                // script_header
                                const _script_headers = [];
                                lodash__default["default"].forEach(_requestPara.header, (item) => {
                                    _script_headers.push(item);
                                });

                                // fix bug for 7.0.8
                                const _script_request_headers_raw = request.formatRequestHeaders(_script_headers, _script_mode);
                                const _script_request_headers = {};

                                if (lodash__default["default"].isPlainObject(_script_request_headers_raw)) {
                                    lodash__default["default"].forEach(_script_request_headers_raw, function (value, key) {
                                        _script_request_headers[mySandbox.replaceIn(key, null, AUTO_CONVERT_FIELD_2_MOCK)] = mySandbox.replaceIn(value, null, AUTO_CONVERT_FIELD_2_MOCK);
                                    });
                                }

                                const _script_header_map = {
                                    urlencoded: 'application/x-www-form-urlencoded',
                                    none: '',
                                    'form-data': 'multipart/form-data',
                                };

                                // script_query
                                const _script_querys = {};
                                if (lodash__default["default"].has(_requestPara, 'query')) {
                                    lodash__default["default"].forEach(request.formatQueries(_requestPara.query), (value, key) => {
                                        // _script_querys[key] = value; // fix bug for 7.0.8
                                        _script_querys[mySandbox.replaceIn(key, null, AUTO_CONVERT_FIELD_2_MOCK)] = mySandbox.replaceIn(value, null, AUTO_CONVERT_FIELD_2_MOCK);
                                    });
                                }

                                // script_body
                                let _script_bodys = {};

                                switch (_script_mode) {
                                    case 'none':
                                        _script_bodys = '';
                                        break;
                                    case 'form-data':
                                    case 'urlencoded':
                                        if (lodash__default["default"].has(_requestPara, 'body') && lodash__default["default"].isArray(_requestPara.body)) {
                                            _requestPara.body.forEach((item) => {
                                                if (parseInt(item.is_checked) > 0) {
                                                    _script_bodys[mySandbox.replaceIn(item.key, null, AUTO_CONVERT_FIELD_2_MOCK)] = mySandbox.replaceIn(item.value, null, AUTO_CONVERT_FIELD_2_MOCK);  // fix bug
                                                }
                                            });
                                        }
                                        break;
                                    case 'json': // fix bug for 7.0.13
                                        if (lodash__default["default"].has(definition, 'request.request.body.raw')) {
                                            _script_bodys = mySandbox.replaceIn(request.formatRawJsonBodys(definition.request.request.body.raw), null, AUTO_CONVERT_FIELD_2_MOCK);
                                        } else {
                                            _script_bodys = '';
                                        }
                                        break;
                                    default: // fix bug for 7.0.13
                                        if (lodash__default["default"].has(definition, 'request.request.body.raw')) {
                                            _script_bodys = mySandbox.replaceIn(definition.request.request.body.raw, null, AUTO_CONVERT_FIELD_2_MOCK);
                                        } else {
                                            _script_bodys = '';
                                        }
                                        break;
                                }

                                // script_request_para
                                // 环境前缀 fix bug
                                // for 多环境
                                // if(_.isString(_.get(temp_env,definition.temp_env)))
                                let _script_pre_url = mySandbox.replaceIn(env_pre_url, null, AUTO_CONVERT_FIELD_2_MOCK);
                                let _script_url = mySandbox.replaceIn(definition.request.url, null, AUTO_CONVERT_FIELD_2_MOCK);

                                // 拼接环境前置URl
                                if (lodash__default["default"].isString(_script_pre_url) && _script_pre_url.length > 0) {
                                    if (!lodash__default["default"].startsWith(lodash__default["default"].toLower(_script_pre_url), 'https://') && !lodash__default["default"].startsWith(lodash__default["default"].toLower(_script_pre_url), 'http://')) {
                                        _script_pre_url = `http://${_script_pre_url}`;
                                    }

                                    // _script_url = urlJoin(_script_pre_url, _script_url);
                                    _script_url = urljoins(_script_pre_url, _script_url);// fix bug for 7.0.8

                                    if (lodash__default["default"].endsWith(_script_pre_url, '/')) { // fix bug
                                        _script_url = lodash__default["default"].replace(_script_url, `${_script_pre_url}:`, `${_script_pre_url.substr(0, _script_pre_url.length - 1)}:`);
                                    } else {
                                        _script_url = lodash__default["default"].replace(_script_url, `${_script_pre_url}/:`, `${_script_pre_url}:`);
                                    }
                                } else if (!lodash__default["default"].startsWith(lodash__default["default"].toLower(_script_url), 'https://') && !lodash__default["default"].startsWith(lodash__default["default"].toLower(_script_url), 'http://')) {
                                    _script_url = `http://${_script_url}`;
                                }

                                const _request_para = {
                                    id: (lodash__default["default"].has(definition, 'request.target_id') ? definition.request.target_id : ''),
                                    name: (lodash__default["default"].has(definition, 'request.name') ? definition.request.name : undefined),
                                    description: (lodash__default["default"].has(definition, 'request.request.description') ? definition.request.request.description : undefined),
                                    url: _script_url,
                                    method: definition.request.method,
                                    timeout: _timeout,
                                    contentType: _script_request_headers['content-type'] ? _script_request_headers['content-type'] : (_script_header_map[_script_mode] ? _script_header_map[_script_mode] : ''),
                                    request_headers: _script_request_headers,
                                    request_querys: _script_querys,
                                    request_bodys: _script_bodys,
                                    data: _script_bodys,
                                    headers: _script_request_headers,
                                };

                                if (!lodash__default["default"].isObject(RUNNER_RESULT_LOG)) {
                                    RUNNER_RESULT_LOG = {};
                                }
                                RUNNER_RESULT_LOG[definition.iteration_id] = {
                                    test_id: definition.test_id,
                                    report_id: RUNNER_REPORT_ID,
                                    parent_id: definition.parent_id,
                                    event_id: definition.event_id,
                                    iteration_id: definition.iteration_id,
                                    type: definition.type,
                                    target_id: definition.target_id,
                                    request: _request_para,
                                    response: {},
                                    http_error: -1,
                                    assert: [],
                                    datetime: dayjs__default["default"]().format('YYYY-MM-DD HH:mm:ss'),
                                };

                                // 执行预执行脚本
                                lodash__default["default"].set(definition, 'script_request', _request_para); // fix bug

                                if (lodash__default["default"].has(_requestPara, 'pre_script') && lodash__default["default"].isString(_requestPara.pre_script)) {
                                    await mySandbox.execute(_requestPara.pre_script, definition, 'pre_script', (err, res) => {
                                        // console.log('pre_script', err, res);
                                        if (err && ignoreError < 1) {
                                            stop(RUNNER_REPORT_ID, String(err));
                                        }
                                    });
                                }

                                let _request = lodash__default["default"].cloneDeep(definition.request);

                                // 替换 _requestPara 的参数变量
                                new Array('header', 'query', 'body', 'resful').forEach((type) => {
                                    _requestPara[type] = lodash__default["default"].values(_requestPara[type]);
                                    _requestPara[type].map((item) => {
                                        if (item.type == 'File') {
                                            lodash__default["default"].assign(item, {
                                                key: mySandbox.replaceIn(item.key, null, AUTO_CONVERT_FIELD_2_MOCK),
                                                value: item.value,
                                            });
                                        } else {
                                            lodash__default["default"].assign(item, {
                                                key: mySandbox.replaceIn(item.key, null, AUTO_CONVERT_FIELD_2_MOCK),
                                                value: mySandbox.replaceIn(item.value, null, AUTO_CONVERT_FIELD_2_MOCK),
                                            });
                                        }
                                    });

                                    if (type == 'body' && lodash__default["default"].has(_request, 'request.body.raw')) {
                                        _request.request.body.raw = mySandbox.replaceIn(_request.request.body.raw, null, AUTO_CONVERT_FIELD_2_MOCK);
                                    }

                                    lodash__default["default"].set(_request, `request.${type}.parameter`, _requestPara[type]);
                                });

                                // 认证 fixed bug
                                if (lodash__default["default"].isObject(_requestPara.auth) && lodash__default["default"].isString(_requestPara.auth.type) && lodash__default["default"].isObject(_requestPara.auth[_requestPara.auth.type])) {
                                    Object.keys(_requestPara.auth[_requestPara.auth.type]).forEach((key) => {
                                        _requestPara.auth[_requestPara.auth.type][key] = mySandbox.replaceIn(_requestPara.auth[_requestPara.auth.type][key], null, AUTO_CONVERT_FIELD_2_MOCK);
                                    });
                                }

                                // 重新渲染请求参数
                                let _target = lodash__default["default"].isObject(RUNNER_RESULT_LOG) ? RUNNER_RESULT_LOG[definition.iteration_id] : {};

                                if (typeof _target === 'object' && lodash__default["default"].isObject(_target.beforeRequest)) {
                                    new Array('query', 'header', 'body').forEach((type) => {
                                        if (lodash__default["default"].has(_request, `request.${type}.parameter`) && lodash__default["default"].isArray(_target.beforeRequest[type])) {
                                            _target.beforeRequest[type].forEach((_item) => {
                                                if (_item.action == 'set') {
                                                    if (lodash__default["default"].isObject(_item.key) || lodash__default["default"].isUndefined(_item.value)) { // 允许直接修改请求体 new features
                                                        if (lodash__default["default"].isArray(_request.request[type].parameter)) {
                                                            _request.request[type].parameter = [];
                                                            if (lodash__default["default"].isObject(_item.key)) {
                                                                lodash__default["default"].forEach(_item.key, (_set_value, _set_key) => {
                                                                    _set_key = lodash__default["default"].trim(_set_key);
                                                                    if (_set_key != '') {
                                                                        _request.request[type].parameter.push({
                                                                            description: '',
                                                                            field_type: 'Text',
                                                                            is_checked: '1',
                                                                            key: mySandbox.replaceIn(_set_key, null, AUTO_CONVERT_FIELD_2_MOCK),
                                                                            not_null: '1',
                                                                            type: 'Text',
                                                                            value: mySandbox.replaceIn(_set_value, null, AUTO_CONVERT_FIELD_2_MOCK),
                                                                        });
                                                                    }
                                                                });
                                                            }
                                                        }
                                                    } else if (lodash__default["default"].isString(_item.key)) {
                                                        const _itemPara = lodash__default["default"].find(_request.request[type].parameter, lodash__default["default"].matchesProperty('key', mySandbox.replaceIn(_item.key, null, AUTO_CONVERT_FIELD_2_MOCK)));

                                                        if (_itemPara) {
                                                            _itemPara.value = mySandbox.replaceIn(_item.value, null, AUTO_CONVERT_FIELD_2_MOCK);
                                                        } else {
                                                            _request.request[type].parameter.push({
                                                                description: '',
                                                                field_type: 'Text',
                                                                is_checked: '1',
                                                                key: mySandbox.replaceIn(_item.key, null, AUTO_CONVERT_FIELD_2_MOCK),
                                                                not_null: '1',
                                                                type: 'Text',
                                                                value: mySandbox.replaceIn(_item.value, null, AUTO_CONVERT_FIELD_2_MOCK),
                                                            });
                                                        }
                                                    }
                                                } else if (_item.action == 'remove') {
                                                    lodash__default["default"].remove(_request.request[type].parameter, lodash__default["default"].matchesProperty('key', mySandbox.replaceIn(_item.key, null, AUTO_CONVERT_FIELD_2_MOCK)));
                                                }
                                            });
                                        }

                                        // fix bug for 7.1.16
                                        if (type == 'header') {
                                            // 重置请求头的 content-type
                                            let _contentType = '';
                                            if (lodash__default["default"].isArray(_target?.beforeRequest?.header)) {
                                                let _t = lodash__default["default"].findLast(_target?.beforeRequest?.header, function (item) {
                                                    return lodash__default["default"].toLower(item.key) == 'content-type' && item.action == 'set'
                                                });
                                                _contentType = lodash__default["default"].isObject(_t) ? lodash__default["default"].toLower(_t.value) : '';
                                            }

                                            let _mode = lodash__default["default"].get(_request, 'request.body.mode');
                                            switch (_contentType) {
                                                case 'application/x-www-form-urlencoded':
                                                    _mode = 'urlencoded';
                                                    break;
                                                case 'application/json':
                                                    _mode = 'json';
                                                    break;
                                                case 'application/javascript':
                                                    _mode = 'javascript';
                                                    break;
                                                case 'application/xml':
                                                    _mode = 'xml';
                                                    break;
                                                case 'application/plain':
                                                case 'text/plain':
                                                    _mode = 'plain';
                                                    break;
                                                case 'text/html':
                                                    _mode = 'html';
                                                    break;
                                            }

                                            if (['urlencoded', 'json', 'javascript', 'xml', 'plain', 'html'].indexOf(_mode) > -1) {
                                                lodash__default["default"].set(_request, 'request.body.mode', _mode);
                                            }
                                        }

                                        // fix bug body 参数为空时，预执行脚本设置body无效的bug for 7.0.13
                                        if (type == 'body' && lodash__default["default"].has(_request, 'request.body.raw') && lodash__default["default"].isArray(_target.beforeRequest.body) && _target.beforeRequest.body.length > 0) {
                                            let _rawParse = null;

                                            _target.beforeRequest[type].forEach((_item) => {
                                                if (_item.action == 'set') {
                                                    if (lodash__default["default"].isObject(_item.key) || lodash__default["default"].isUndefined(_item.value)) { // 允许直接修改请求体 new features
                                                        if (lodash__default["default"].isObject(_item.key)) {
                                                            _request.request.body.raw = mySandbox.replaceIn(jsonBigint__default["default"].stringify(_item.key), null, AUTO_CONVERT_FIELD_2_MOCK);
                                                        } else if (lodash__default["default"].isString(_item.key)) {
                                                            _request.request.body.raw = mySandbox.replaceIn(String(_item.key), null, AUTO_CONVERT_FIELD_2_MOCK); // fix bug
                                                        } else if (lodash__default["default"].isNumber(_item.key)) {
                                                            _request.request.body.raw = String(_item.key);
                                                        }
                                                    } else if (lodash__default["default"].isString(_item.key) && apipostTools__default["default"].isJson5(_request.request.body.raw)) {
                                                        try {
                                                            _rawParse = jsonBigint__default["default"].parse(stripJsonComments__default["default"](_request.request.body.raw));
                                                        } catch (e) {
                                                            _rawParse = json5__default["default"].parse(_request.request.body.raw);
                                                        }
                                                        lodash__default["default"].set(_rawParse, mySandbox.replaceIn(_item.key, null, AUTO_CONVERT_FIELD_2_MOCK), mySandbox.replaceIn(_item.value, null, AUTO_CONVERT_FIELD_2_MOCK));

                                                        if (lodash__default["default"].isObject(_rawParse)) {
                                                            _request.request.body.raw = jsonBigint__default["default"].stringify(_rawParse);
                                                        } else {
                                                            _request.request.body.raw = _rawParse;
                                                        }
                                                    }
                                                } else if (_item.action == 'remove' && apipostTools__default["default"].isJson5(_request.request.body.raw)) {
                                                    try {
                                                        _rawParse = jsonBigint__default["default"].parse(stripJsonComments__default["default"](_request.request.body.raw));
                                                    } catch (e) {
                                                        _rawParse = json5__default["default"].parse(_request.request.body.raw);
                                                    }
                                                    lodash__default["default"].unset(_rawParse, mySandbox.replaceIn(_item.key, null, AUTO_CONVERT_FIELD_2_MOCK));

                                                    if (lodash__default["default"].isObject(_rawParse)) {
                                                        _request.request.body.raw = jsonBigint__default["default"].stringify(_rawParse);
                                                    } else {
                                                        _request.request.body.raw = _rawParse;
                                                    }
                                                }
                                            });
                                        }
                                    });
                                }

                                if (lodash__default["default"].isObject(_requestPara.auth[_requestPara.auth.type])) {
                                    _requestPara.auth[_requestPara.auth.type] = lodash__default["default"].mapValues(_requestPara.auth[_requestPara.auth.type], val => mySandbox.replaceIn(val, null, AUTO_CONVERT_FIELD_2_MOCK));
                                    // console.log(_request, _requestPara);
                                    lodash__default["default"].set(_request, 'request.auth.type', _requestPara.auth.type); // fix bug
                                    lodash__default["default"].set(_request, `request.auth.${_requestPara.auth.type}`, _requestPara.auth[_requestPara.auth.type]);
                                }

                                // url 兼容
                                let _url = _request.request.url ? _request.request.url : _request.url;
                                _url = mySandbox.replaceIn(_url, null, AUTO_CONVERT_FIELD_2_MOCK);

                                // fixed bug add 替换路径变量
                                if (lodash__default["default"].isArray(_requestPara.resful) && _requestPara.resful.length > 0) {
                                    _requestPara.resful.forEach((_resful) => {
                                        _resful.key = lodash__default["default"].trim(_resful.key);

                                        if (_resful.is_checked > 0 && _resful.key !== '') {
                                            _url = lodash__default["default"].replace(_url, `:${_resful.key}`, _resful.value);
                                            _url = lodash__default["default"].replace(_url, `{${_resful.key}}`, _resful.value);
                                        }
                                    });
                                }

                                // 环境前缀 fix bug
                                let _pre_url = mySandbox.replaceIn(env_pre_url, null, AUTO_CONVERT_FIELD_2_MOCK);

                                // 拼接环境前置URl
                                if (lodash__default["default"].isString(_pre_url) && _pre_url.length > 0) {
                                    if (!lodash__default["default"].startsWith(lodash__default["default"].toLower(_pre_url), 'https://') && !lodash__default["default"].startsWith(lodash__default["default"].toLower(_pre_url), 'http://')) {
                                        _pre_url = `http://${_pre_url}`;
                                    }

                                    // _url = urlJoin(_pre_url, _url);
                                    _url = urljoins(_pre_url, _url); // fix bug for 7.0.8

                                    if (lodash__default["default"].endsWith(_pre_url, '/')) { // fix bug
                                        _url = lodash__default["default"].replace(_url, `${_pre_url}:`, `${_pre_url.substr(0, _pre_url.length - 1)}:`);
                                    } else {
                                        _url = lodash__default["default"].replace(_url, `${_pre_url}/:`, `${_pre_url}:`);
                                    }
                                } else if (!lodash__default["default"].startsWith(lodash__default["default"].toLower(_url), 'https://') && !lodash__default["default"].startsWith(lodash__default["default"].toLower(_url), 'http://')) {
                                    _url = `http://${_url}`;
                                }
                                // _url=encodeURI(_url);
                                lodash__default["default"].set(_request, 'url', _url);
                                lodash__default["default"].set(_request, 'request.url', _url);

                                let _isHttpError = -1;

                                // cookie
                                // 已修复 cookie 无法使用的问题
                                if (typeof cookies === 'object' && lodash__default["default"].has(cookies, 'switch') && lodash__default["default"].has(cookies, 'data')) {
                                    if (cookies.switch > 0 && lodash__default["default"].isArray(cookies.data)) {
                                        const _cookieArr = [];
                                        cookies.data.forEach((_cookie) => {
                                            if (typeof _cookie.name === 'undefined' && typeof _cookie.key === 'string') {
                                                _cookie.name = _cookie.key;
                                            }
                                            const cookieStr = checkValidCookie__default["default"].isvalid(_url, _cookie);

                                            if (cookieStr) {
                                                _cookieArr.push(cookieStr.cookie);
                                            }
                                        });

                                        if (_cookieArr.length > 0) {
                                            if (lodash__default["default"].has(_request, 'request.header.parameter')) {
                                                const _targetHeaderCookie = lodash__default["default"].find(_request.request.header.parameter, o => lodash__default["default"].trim(lodash__default["default"].toLower(o.key)) == 'cookie');

                                                if (_targetHeaderCookie && _targetHeaderCookie.is_checked > 0) {
                                                    _targetHeaderCookie.value = `${_cookieArr.join(';')};${_targetHeaderCookie.value}`; // fix bug for 7.0.8
                                                } else {
                                                    _request.request.header.parameter.push({
                                                        key: 'cookie',
                                                        value: _cookieArr.join(';'), // fix cookie bug
                                                        description: '',
                                                        not_null: 1,
                                                        field_type: 'String',
                                                        type: 'Text',
                                                        is_checked: 1,
                                                    });
                                                }
                                            } else {
                                                lodash__default["default"].set(_request, 'request.header.parameter', [{
                                                    key: 'cookie',
                                                    value: _cookieArr.join(';'), // fix cookie bug
                                                    description: '',
                                                    not_null: 1,
                                                    field_type: 'String',
                                                    type: 'Text',
                                                    is_checked: 1,
                                                }]);
                                            }
                                        }
                                    }
                                }

                                //如果是mock环境，则携带apipost_id
                                if (`${option.env_id}` === '-2') {
                                    //判断query数组内是否包含apipost_id
                                    const requestApipostId = _request?.request?.query?.parameter.find(item => item.key === 'apipost_id');
                                    if (lodash__default["default"].isUndefined(requestApipostId)) {
                                        _request.request.query.parameter.push({
                                            key: 'apipost_id',
                                            value: _request.target_id.substr(0, 6),
                                            description: '',
                                            not_null: 1,
                                            field_type: 'String',
                                            type: 'Text',
                                            is_checked: 1,
                                        });
                                    }
                                }

                                try {
                                    // console.log(_request)
                                    // 合并请求参数
                                    res = await request.request(_request);
                                } catch (e) {
                                    res = e;
                                }

                                if (res.status === 'error') {
                                    _isHttpError = 1;
                                    RUNNER_ERROR_COUNT++;

                                    if (scene == 'auto_test') {
                                        cliConsole(`\n${_request.method} ${_request.url}`.grey);
                                        cliConsole(`\t${RUNNER_ERROR_COUNT}. HTTP 请求失败`.bold.red); // underline.
                                    }
                                } else {
                                    _isHttpError = -1;
                                    if (scene == 'auto_test') {
                                        cliConsole(`\n${_request.method} ${_request.url} [${res.data.response.code} ${res.data.response.status}, ${res.data.response.responseSize}KB, ${res.data.response.responseTime}ms]`.grey);
                                        cliConsole('\t✓'.green + ' HTTP 请求成功'.grey);
                                    }
                                }

                                // 优化返回体结构
                                let _response = lodash__default["default"].cloneDeep(res);

                                if (_response.status == 'success' && lodash__default["default"].isObject(_response.data.response)) {
                                    lodash__default["default"].unset(_response, 'data.response.base64Body');
                                    lodash__default["default"].unset(_response, 'data.response.header');
                                    lodash__default["default"].unset(_response, 'data.response.resHeaders');
                                    lodash__default["default"].unset(_response, 'data.response.json');
                                    lodash__default["default"].unset(_response, 'data.response.raw');
                                    lodash__default["default"].unset(_response, 'data.response.rawBody');
                                    lodash__default["default"].unset(_response, 'data.response.rawCookies');
                                    lodash__default["default"].unset(_response, 'data.response.cookies');
                                }

                                try {
                                    lodash__default["default"].set(_response, 'data.response.stream.data', Array.from(zlib__default["default"].deflateSync(Buffer.from(_response.data.response.stream.data))));
                                } catch (err) { } // 此错误无需中止运行

                                lodash__default["default"].assign(_target, {
                                    request: _request,
                                    response: _response,
                                    http_error: _isHttpError,
                                });

                                // fix bug for 7.1.16
                                if (lodash__default["default"].isArray(_response?.data?.response?.resCookies)) {
                                    lodash__default["default"].forEach(_response?.data?.response?.resCookies, function (item) {
                                        if (lodash__default["default"].isArray(cookies?.data)) {
                                            let _cookieItemArray = lodash__default["default"].remove(cookies?.data, function (c) {
                                                return c.name == item.name && c.domain == item.domain && c.path == item.path;
                                            });

                                            if (_cookieItemArray.length == 0) {
                                                item.cookie_id = uuid__default["default"].v4();
                                                item.project_id = _request.project_id;
                                                cookies.data.push(item);
                                            } else {
                                                lodash__default["default"].forEach(_cookieItemArray, function (_cookieItem) {
                                                    item.cookie_id = _cookieItem.cookie_id;
                                                    item.project_id = _cookieItem.project_id;
                                                    cookies.data.push(item);
                                                });
                                            }
                                        }
                                    });
                                }

                                // 发送console
                                // 修改请求url
                                const requestUrl = request.setQueryString(_request.request.url, request.formatQueries(_request.request.query.parameter)).uri;

                                if (scene != 'auto_test') { // / done
                                    if (res.status === 'error') {
                                        let _formPara = {};

                                        if (lodash__default["default"].indexOf(['form-data', 'urlencoded'], _request.request.body.mode)) {
                                            if (lodash__default["default"].has(_request, 'request.body.parameter')) {
                                                _request.request.body.parameter.forEach((_para) => {
                                                    if (_para.is_checked > 0) {
                                                        if (!_formPara[_para.key]) {
                                                            _formPara[_para.key] = [];
                                                        }
                                                        _formPara[_para.key].push(_para.value);
                                                    }
                                                });
                                            }
                                        } else {
                                            _formPara = _request.request.body.raw;
                                        }

                                        // 请求控制台信息
                                        emitRuntimeEvent({
                                            action: 'console',
                                            method: 'request',
                                            message: {
                                                data: {
                                                    request: {
                                                        id: (lodash__default["default"].has(definition, 'request.target_id') ? definition.request.target_id : ''),
                                                        name: (lodash__default["default"].has(definition, 'request.name') ? definition.request.name : undefined),
                                                        description: (lodash__default["default"].has(definition, 'request.request.description') ? definition.request.request.description : undefined),
                                                        method: _request.method,
                                                        url: requestUrl,
                                                        request_bodys: lodash__default["default"].indexOf(['form-data', 'urlencoded'], _request.request.body.mode) ? lodash__default["default"].mapValues(_formPara, o => lodash__default["default"].size(o) > 1 ? o : o[0]) : _formPara,
                                                        request_headers: {
                                                            ...request.formatRequestHeaders(_request.request.header.parameter),
                                                            ...request.createAuthHeaders(_request),
                                                        },
                                                        data: lodash__default["default"].indexOf(['form-data', 'urlencoded'], _request.request.body.mode) ? lodash__default["default"].mapValues(_formPara, o => lodash__default["default"].size(o) > 1 ? o : o[0]) : _formPara,
                                                        headers: {
                                                            ...request.formatRequestHeaders(_request.request.header.parameter),
                                                            ...request.createAuthHeaders(_request),
                                                        },
                                                    },
                                                    response: {},
                                                    message: _response.message,
                                                    status: 'error',
                                                },
                                            },
                                            timestamp: Date.now(),
                                            datetime: dayjs__default["default"]().format('YYYY-MM-DD HH:mm:ss'),
                                        });
                                    } else {

                                        lodash__default["default"].set(_response, 'data.request.url', requestUrl);
                                        emitRuntimeEvent({
                                            action: 'console',
                                            method: 'request',
                                            message: _response,
                                            timestamp: Date.now(),
                                            datetime: dayjs__default["default"]().format('YYYY-MM-DD HH:mm:ss'),
                                        });
                                    }
                                }

                                // 执行后执行脚本
                                if (lodash__default["default"].has(_requestPara, 'test') && lodash__default["default"].isString(_requestPara.test)) {
                                    if (lodash__default["default"].isString(_global_asserts_script) && lodash__default["default"].trim(_global_asserts_script) != '') {
                                        _requestPara.test = `${_requestPara.test}\r\n${_global_asserts_script}`;
                                    }

                                    await mySandbox.execute(_requestPara.test, lodash__default["default"].assign(definition, { response: res }), 'test', (err, exec_res) => {
                                        if (err && ignoreError < 1) {
                                            stop(RUNNER_REPORT_ID, String(err));
                                        } else if (lodash__default["default"].has(exec_res, 'raw.responseText') && lodash__default["default"].has(res, 'data.response.raw.responseText') && exec_res.raw.responseText != res.data.response.raw.responseText) {
                                            lodash__default["default"].set(_response, 'data.response.changeBody', exec_res.raw.responseText);
                                        }
                                    });
                                }

                                // fit support response && request
                                if (lodash__default["default"].isObject(_request_para)) {
                                    lodash__default["default"].set(PREV_REQUEST, 'request', _request_para);
                                }

                                if (lodash__default["default"].has(res, 'data.response')) {
                                    lodash__default["default"].set(PREV_REQUEST, 'response', res.data.response);
                                }
                                // fix bug
                                if (definition.event_id != '0' && scene == 'auto_test') {
                                    if (lodash__default["default"].find(_target.assert, lodash__default["default"].matchesProperty('status', 'error'))) {
                                        _target.assert_error = 1;
                                    } else {
                                        _target.assert_error = -1;
                                    }

                                    emitRuntimeEvent({
                                        action: 'current_event_id',
                                        combined_id,
                                        test_id: definition.test_id,
                                        current_event_id: definition.event_id,
                                        test_log: _target,
                                    });
                                }

                                _requestPara = _request = _response = res = _parent_ids = _target = null;
                            }
                            break;
                        case 'for':
                            if (lodash__default["default"].isArray(definition.children) && definition.children.length > 0) {
                                let _for_option = lodash__default["default"].cloneDeep(option);

                                if (lodash__default["default"].get(definition, 'condition.enable_data') > 0 && lodash__default["default"].isArray(lodash__default["default"].get(definition, 'condition.iterationData'))) {
                                    _for_option.iterationData = definition.condition.iterationData;
                                }

                                for (let i = 0; i < mySandbox.replaceIn(definition.condition.limit); i++) {
                                    await run(definition.children, lodash__default["default"].assign(_for_option, { sleep: parseInt(definition.condition.sleep) }), initFlag + 1, i);
                                }
                            }
                            break;
                        case 'while':
                            if (lodash__default["default"].isArray(definition.children) && definition.children.length > 0) {
                                let _while_option = lodash__default["default"].cloneDeep(option);

                                if (lodash__default["default"].get(definition, 'condition.enable_data') > 0 && lodash__default["default"].isArray(lodash__default["default"].get(definition, 'condition.iterationData'))) {
                                    _while_option.iterationData = definition.condition.iterationData;
                                }

                                const end = Date.now() + parseInt(definition.condition.timeout);
                                lodash__default["default"].set(definition, 'runtime.condition', `${mySandbox.replaceIn(definition.condition.var)} ${definition.condition.compare} ${mySandbox.replaceIn(definition.condition.value)}`);

                                let _i = 0;
                                while ((returnBoolean(mySandbox.replaceIn(definition.condition.var), definition.condition.compare, mySandbox.replaceIn(definition.condition.value)))) {
                                    if (Date.now() > end) {
                                        break;
                                    }

                                    await run(definition.children, lodash__default["default"].assign(_while_option, { sleep: parseInt(definition.condition.sleep) }), initFlag + 1, _i);
                                    _i++;
                                }
                            }
                            break;
                        case 'begin':
                            let _begin_option = lodash__default["default"].cloneDeep(option);

                            if (lodash__default["default"].get(definition, 'condition.enable_data') > 0 && lodash__default["default"].isArray(lodash__default["default"].get(definition, 'condition.iterationData'))) {
                                _begin_option.iterationData = definition.condition.iterationData;
                            }
                            await run(definition.children, _begin_option, initFlag + 1, loopCount);
                            break;
                    }

                    if (definition.type != 'api' && definition.type != 'sample' && definition.event_id != '0' && scene == 'auto_test') {
                        emitRuntimeEvent({
                            action: 'current_event_id',
                            combined_id,
                            test_id: definition.test_id,
                            current_event_id: definition.event_id,
                            test_log: null, // 非 api 不传 test_log
                        });
                    }

                    if (initFlag <= 1) {
                        RUNNER_RUNTIME_POINTER++;
                    }

                    // 进度条
                    if (RUNNER_TOTAL_COUNT >= RUNNER_RUNTIME_POINTER && scene == 'auto_test') {
                        RUNNER_PROGRESS = lodash__default["default"].floor(lodash__default["default"].divide(RUNNER_RUNTIME_POINTER, RUNNER_TOTAL_COUNT), 2);

                        emitRuntimeEvent({
                            action: 'progress',
                            progress: RUNNER_PROGRESS,
                            combined_id,
                            test_id: definition.test_id,
                            current_event_id: definition.event_id,
                        });
                    }

                    // 完成
                    if (RUNNER_TOTAL_COUNT == RUNNER_RUNTIME_POINTER) {
                        if (scene == 'auto_test') { // 自动化测试
                            // 获取未跑的 event
                            let ignoreEvents = [];
                            (function getIgnoreAllApis(initDefinitions) {
                                if (!lodash__default["default"].isArray(initDefinitions));
                                // fix bug for 7.0.8
                                if (lodash__default["default"].isArray(initDefinitions)) {
                                    initDefinitions.forEach((item) => {
                                        if ((item.type == 'api' || item.type == 'sample') && !lodash__default["default"].find(RUNNER_RESULT_LOG, lodash__default["default"].matchesProperty('event_id', item.event_id))) {
                                            const _iteration_id = uuid__default["default"].v4();

                                            if (lodash__default["default"].isObject(RUNNER_RESULT_LOG)) {
                                                RUNNER_RESULT_LOG[_iteration_id] = {
                                                    test_id: item.test_id,
                                                    report_id: RUNNER_REPORT_ID,
                                                    parent_id: item.parent_id,
                                                    event_id: item.event_id,
                                                    iteration_id: _iteration_id,
                                                    type: item.type,
                                                    target_id: item.target_id,
                                                    request: item.request,
                                                    response: {},
                                                    http_error: -2,
                                                    assert: [],
                                                    datetime: dayjs__default["default"]().format('YYYY-MM-DD HH:mm:ss'),
                                                };

                                                ignoreEvents.push(RUNNER_RESULT_LOG[_iteration_id]);
                                            }
                                        }

                                        if (lodash__default["default"].isArray(item.children)) {
                                            getIgnoreAllApis(item.children);
                                        }
                                    });
                                }
                            }(initDefinitions));

                            const _runReport = calculateRuntimeReport(RUNNER_RESULT_LOG, initDefinitions, RUNNER_REPORT_ID, { combined_id, test_events, default_report_name, user, env_name, env });
                            _runReport.logList = Object.values(RUNNER_RESULT_LOG)
                            emitRuntimeEvent({
                                action: 'complate',
                                combined_id,
                                ignore_error,
                                enable_sandbox,
                                envs: {
                                    globals: mySandbox.variablesScope.globals,
                                    environment: lodash__default["default"].assign(mySandbox.variablesScope.environment, mySandbox.variablesScope.variables), // fix variables bug
                                },
                                ignore_events: ignoreEvents,
                                test_report: _runReport,
                            });

                            // 打印报告
                            const reportTable = new cliTable3__default["default"]({
                                style: { padding: 5, head: [], border: [] },
                            });

                            // fix bug for 7.1.16
                            reportTable.push(
                                [{ content: 'The result of API test', colSpan: 4, hAlign: 'center' }],
                                ['', { content: 'passed', hAlign: 'center' }, { content: 'failed', hAlign: 'center' }, { content: 'ignore', hAlign: 'center' }],
                                [{ content: 'request', hAlign: 'left' }, { content: `${_runReport.http.passed}`, hAlign: 'center' }, { content: `${_runReport.http.failure}`, hAlign: 'center' }, { content: `${_runReport.ignore_count}`, rowSpan: 2, hAlign: 'center', vAlign: 'center' }],
                                [{ content: 'assertion', hAlign: 'left' }, { content: `${_runReport.assert.passed}`, hAlign: 'center' }, { content: `${_runReport.assert.failure}`, hAlign: 'center' }],
                                [{ content: `total number of api: ${_runReport.total_count}, ignore: ${_runReport.ignore_count}`, colSpan: 4, hAlign: 'left' }],
                                [{ content: `total data received: ${_runReport.total_received_data} KB (approx)`, colSpan: 4, hAlign: 'left' }],
                                [{ content: `total response time: ${_runReport.total_response_time} 毫秒, average response time: ${_runReport.average_response_time} 毫秒`, colSpan: 4, hAlign: 'left' }],
                                [{ content: `total run duration: ${_runReport.long_time}`, colSpan: 4, hAlign: 'left' }],
                                [{ content: 'Generated by apipost-cli ( https://github.com/Apipost-Team/apipost-cli )', colSpan: 4, hAlign: 'center' }],
                            );

                            cliConsole(reportTable.toString());


                            let cliCounter = 0;

                            if (lodash__default["default"].size(_runReport.http_errors) > 0) {
                                const httpFailedTable = new cliTable3({
                                    chars: {
                                        top: '',
                                        'top-mid': '',
                                        'top-left': '',
                                        'top-right': '',
                                        bottom: '',
                                        'bottom-mid': '',
                                        'bottom-left': '',
                                        'bottom-right': '',
                                        left: '',
                                        'left-mid': '',
                                        mid: '',
                                        'mid-mid': '',
                                        right: '',
                                        'right-mid': '',
                                        middle: ' ',
                                    },
                                    style: { padding: 5, head: [], border: [] },
                                });

                                // fix bug for 7.1.16
                                httpFailedTable.push(
                                    [{ content: '', colSpan: 2 }],
                                    [{ content: '#', hAlign: 'center' }, { content: 'failure', hAlign: 'left' }, { content: 'detail', hAlign: 'left' }],
                                ); // fix bug for 7.0.8 bug

                                lodash__default["default"].forEach(_runReport.http_errors, (item) => {
                                    cliCounter++;
                                    httpFailedTable.push(
                                        [{ content: '', colSpan: 2 }],
                                        [{ content: `${cliCounter}.`, hAlign: 'center' }, { content: '请求错误', hAlign: 'left' }, { content: `${`${lodash__default["default"].get(item, 'response.status')}` + '\t' + `${lodash__default["default"].get(item, 'request.url')}` + '\n'}${`${lodash__default["default"].get(item, 'response.message')}`}`, hAlign: 'left' }],
                                    );
                                });

                                cliConsole(httpFailedTable.toString());
                            }

                            if (lodash__default["default"].size(_runReport.assert_errors) > 0) {
                                const failedTable = new cliTable3({
                                    chars: {
                                        top: '',
                                        'top-mid': '',
                                        'top-left': '',
                                        'top-right': '',
                                        bottom: '',
                                        'bottom-mid': '',
                                        'bottom-left': '',
                                        'bottom-right': '',
                                        left: '',
                                        'left-mid': '',
                                        mid: '',
                                        'mid-mid': '',
                                        right: '',
                                        'right-mid': '',
                                        middle: ' ',
                                    },
                                    style: { padding: 5, head: [], border: [] },
                                });

                                // fix bug for 7.1.16
                                failedTable.push(
                                    [{ content: '', colSpan: 2 }],
                                    [{ content: '#', hAlign: 'center' }, { content: 'failure', hAlign: 'left' }, { content: 'detail', hAlign: 'left' }],
                                ); // fix bug for 7.0.8 bug

                                lodash__default["default"].forEach(_runReport.assert_errors, (item) => {
                                    lodash__default["default"].forEach(item.assert, (assert) => {
                                        cliCounter++;
                                        failedTable.push(
                                            [{ content: '', colSpan: 2 }],
                                            [{ content: `${cliCounter}.`, hAlign: 'center' }, { content: '断言错误', hAlign: 'left' }, { content: `${`${assert.expect}` + '\n'}${`${assert.result}`}`, hAlign: 'left' }],
                                        );
                                    });
                                });

                                cliConsole(failedTable.toString());
                            }

                            ignoreEvents = null;
                        } else { // 接口请求
                            const _http = lodash__default["default"].isObject(RUNNER_RESULT_LOG) ? RUNNER_RESULT_LOG[definition.iteration_id] : {};

                            emitRuntimeEvent({
                                action: 'http_complate',
                                envs: {
                                    globals: mySandbox.variablesScope.globals,
                                    environment: lodash__default["default"].assign(mySandbox.variablesScope.environment, mySandbox.variablesScope.variables), // fix variables bug
                                },
                                data: {
                                    script_error: _http.script_error, // fixed script error bug
                                    visualizer_html: _http.visualizer_html, // fixed 可视化 bug
                                    assert: _http.assert,
                                    target_id: _http.target_id,
                                    response: _http.response,
                                },
                                timestamp: Date.now(),
                                datetime: dayjs__default["default"]().format('YYYY-MM-DD HH:mm:ss'),
                            });
                        }

                        runInit();
                        RUNNER_RESULT_LOG = initDefinitions = null;

                        if (RUNNER_STOP[RUNNER_REPORT_ID]) {
                            delete RUNNER_STOP[RUNNER_REPORT_ID];
                        }
                    }
                }
            }
        }
    }

    // 构造一个执行对象
    Object.defineProperty(this, 'run', {
        value: run,
    });

    Object.defineProperty(this, 'stop', {
        value: stop,
    });
};

var Runtime_1 = Runtime;
var Collection_1 = Collection;

var runtime = {
    Runtime: Runtime_1,
    Collection: Collection_1
};

exports.Collection = Collection_1;
exports.Runtime = Runtime_1;
exports["default"] = runtime;
